diff --git a/arch/i386/Makefile b/arch/i386/Makefile
index 425361fd9..2fa008fa7 100644
--- a/arch/i386/Makefile
+++ b/arch/i386/Makefile
@@ -11,3 +11,11 @@ ifeq ($(CONFIG_STACK_OPTIMIZATION_386),y)
 # from aligning stack to 16 bytes. (Which is gcc's way of supporting SSE).
 CFLAGS += $(call cc-option,-mpreferred-stack-boundary=2,)
 endif
+
+# "Control how GCC aligns variables.
+# Supported values for type are compat uses increased alignment value
+# compatible uses GCC 4.8 and earlier, abi uses alignment value as specified by the psABI,
+# and cacheline uses increased alignment value to match the cache line size.
+# compat is the default."
+# "abi" seems to be somewhat successful in preventing oversealous data alignment.
+CFLAGS += $(call cc-option,-malign-data=abi,)
