diff --color -Naur linux-4.19.132-orig/Makefile linux-4.19.132/Makefile
--- linux-4.19.132-orig/Makefile	2020-07-13 11:34:26.368791506 +0300
+++ linux-4.19.132/Makefile	2020-07-13 11:35:04.177300306 +0300
@@ -653,9 +653,9 @@
 KBUILD_CFLAGS	+= $(call cc-disable-warning, address-of-packed-member)
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
-KBUILD_CFLAGS   += -Os
+KBUILD_CFLAGS   += -Os $(EXTRA_OPTIMIZATION)
 else
-KBUILD_CFLAGS   += -O2
+KBUILD_CFLAGS   += -O2 $(EXTRA_OPTIMIZATION)
 endif
 
 # Tell gcc to never replace conditional load with a non-conditional one
