diff --color -Naur linux-4.19.164-orig/net/ipv4/netfilter/ip_tables.c linux-4.19.164/net/ipv4/netfilter/ip_tables.c
--- linux-4.19.164-orig/net/ipv4/netfilter/ip_tables.c	2020-12-30 21:43:14.428632076 +0300
+++ linux-4.19.164/net/ipv4/netfilter/ip_tables.c	2020-12-30 21:51:01.244276419 +0300
@@ -223,6 +223,33 @@
 	return (void *)entry + entry->next_offset;
 }
 
+static bool
+ipt_handle_default_rule(struct ipt_entry *e, unsigned int *verdict)
+{
+	struct xt_entry_target *t;
+	struct xt_standard_target *st;
+
+	if (e->target_offset != sizeof(struct ipt_entry))
+		return false;
+
+	if (!(e->ip.flags & IPT_F_NO_DEF_MATCH))
+		return false;
+
+	t = ipt_get_target(e);
+	if (t->u.kernel.target->target)
+		return false;
+
+	st = (struct xt_standard_target *) t;
+	if (st->verdict == XT_RETURN)
+		return false;
+
+	if (st->verdict >= 0)
+		return false;
+
+	*verdict = (unsigned)(-st->verdict) - 1;
+	return true;
+}
+
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ipt_do_table(struct sk_buff *skb,
