diff -Naur iproute2-4.12.0-orig/include/linux/if_ether.h iproute2-4.12.0/include/linux/if_ether.h
--- iproute2-4.12.0-orig/include/linux/if_ether.h	2017-08-18 15:35:35.681981626 +0300
+++ iproute2-4.12.0/include/linux/if_ether.h	2017-08-18 15:36:59.298622722 +0300
@@ -22,6 +22,7 @@
 #define _LINUX_IF_ETHER_H
 
 #include <linux/types.h>
+#include <linux/libc-compat.h>
 
 /*
  *	IEEE 802.3 Ethernet magic constants.  The frame sizes omit the preamble
@@ -142,11 +143,13 @@
  *	This is an Ethernet frame header.
  */
 
+#if __UAPI_DEF_ETHHDR
 struct ethhdr {
 	unsigned char	h_dest[ETH_ALEN];	/* destination eth addr	*/
 	unsigned char	h_source[ETH_ALEN];	/* source ether addr	*/
 	__be16		h_proto;		/* packet type ID field	*/
 } __attribute__((packed));
+#endif
 
 
 #endif /* _LINUX_IF_ETHER_H */
