diff --git a/zlib.pc.cmakein b/zlib.pc.cmakein
index 7744dc2..5ed025b 100644
--- a/zlib.pc.cmakein
+++ b/zlib.pc.cmakein
@@ -1,8 +1,8 @@
 prefix=@CMAKE_INSTALL_PREFIX@
 exec_prefix=${prefix}
-libdir=@PC_INSTALL_LIB_DIR@
-sharedlibdir=${libdir}
-includedir=@PC_INSTALL_INC_DIR@
+libdir=${exec_prefix}/lib
+sharedlibdir=${exec_prefix}/lib
+includedir=${prefix}/include
 
 Name: zlib@SUFFIX@
 Description: zlib-ng compression library
