From f78641245ba1f729f03bca8a9403e835bc96782e Mon Sep 17 00:00:00 2001
From: Joachim Nilsson <troglobit@gmail.com>
Date: Thu, 6 Feb 2020 05:50:05 +0100
Subject: [PATCH] Fix #17: Missing freeifaddrs() causing major memory
 leak

Only in Linux backend.

Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
Signed-off-by: Alexander Sverdlin <alexander.sverdlin@gmail.com>
---
 linux.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/linux.c b/linux.c
index ad5714c..a657912 100644
--- a/linux.c
+++ b/linux.c
@@ -292,6 +292,7 @@ void get_netinfo(netinfo_t *netinfo)
 	}
 
 	parse_file("/proc/net/dev", fields, NELEMS(fields), 0);
+	freeifaddrs(ifap);
 }
 
 #endif /* __linux__ */
-- 
2.23.0

