diff --color -Naur mc-4.8.25-orig/lib/tty/tty.c mc-4.8.25/lib/tty/tty.c
--- mc-4.8.25-orig/lib/tty/tty.c	2020-07-20 12:27:08.591727297 +0300
+++ mc-4.8.25/lib/tty/tty.c	2020-07-20 12:36:46.532551003 +0300
@@ -370,7 +370,7 @@
     if (xmouse_seq != NULL)
     {
         if (strcmp (xmouse_seq, ESC_STR "[<") == 0)
-            xmouse_seq = ESC_STR "[M";
+            xmouse_seq = NULL;
 
         xmouse_extended_seq = ESC_STR "[<";
     }
