diff --git a/Makefile b/Makefile
index 24f5955c90..f6cd169f0e 100644
--- a/Makefile
+++ b/Makefile
@@ -86,12 +86,6 @@ endif
 
 # check prerequisites before starting to build
 prereq: $(target/stamp-prereq) tmp/.prereq_packages
-	@if [ ! -f "$(INCLUDE_DIR)/site/$(ARCH)" ]; then \
-		echo 'ERROR: Missing site config for architecture "$(ARCH)" !'; \
-		echo '       The missing file will cause configure scripts to fail during compilation.'; \
-		echo '       Please provide a "$(INCLUDE_DIR)/site/$(ARCH)" file and restart the build.'; \
-		exit 1; \
-	fi
 
 $(BIN_DIR)/profiles.json: FORCE
 	$(if $(CONFIG_JSON_OVERVIEW_IMAGE_INFO), \
diff --git a/include/package.mk b/include/package.mk
index 50bd838180..1be951bfcf 100644
--- a/include/package.mk
+++ b/include/package.mk
@@ -119,7 +119,7 @@ include $(INCLUDE_DIR)/autotools.mk
 _pkg_target:=$(if $(QUILT),,.)
 
 override MAKEFLAGS=
-CONFIG_SITE:=$(INCLUDE_DIR)/site/$(ARCH)
+CONFIG_SITE:=$(TOOLCHAIN_DIR)/config.site
 CUR_MAKEFILE:=$(filter-out Makefile,$(firstword $(MAKEFILE_LIST)))
 SUBMAKE:=$(NO_TRACE_MAKE) $(if $(CUR_MAKEFILE),-f $(CUR_MAKEFILE))
 PKG_CONFIG_PATH=$(STAGING_DIR)/usr/lib/pkgconfig:$(STAGING_DIR)/usr/share/pkgconfig
diff --git a/include/site/aarch64 b/include/site/aarch64
deleted file mode 100644
index c5aa9c5543..0000000000
--- a/include/site/aarch64
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=8
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=8
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=8
-ac_cv_sizeof_ssize_t=8
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=8
diff --git a/include/site/aarch64_be b/include/site/aarch64_be
deleted file mode 100644
index 19e75ab991..0000000000
--- a/include/site/aarch64_be
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=8
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=8
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=8
-ac_cv_sizeof_ssize_t=8
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=8
diff --git a/include/site/arc b/include/site/arc
deleted file mode 100644
index 72a3805c64..0000000000
--- a/include/site/arc
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/arm b/include/site/arm
deleted file mode 100644
index 72a3805c64..0000000000
--- a/include/site/arm
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/armeb b/include/site/armeb
deleted file mode 100644
index a5626a722b..0000000000
--- a/include/site/armeb
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/i386 b/include/site/i386
deleted file mode 100644
index 78f1557b79..0000000000
--- a/include/site/i386
+++ /dev/null
@@ -1,3 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/i486
-
diff --git a/include/site/i486 b/include/site/i486
deleted file mode 100644
index 72a3805c64..0000000000
--- a/include/site/i486
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/i686 b/include/site/i686
deleted file mode 100644
index 78f1557b79..0000000000
--- a/include/site/i686
+++ /dev/null
@@ -1,3 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/i486
-
diff --git a/include/site/linux b/include/site/linux
deleted file mode 100644
index b193d25c45..0000000000
--- a/include/site/linux
+++ /dev/null
@@ -1,78 +0,0 @@
-ac_atomic_add=yes
-ac_atomic_sub=yes
-ac_cv_c_gettext_without_libintl=yes
-ac_cv_c_long_double=no
-ac_cv_conv_longlong_to_float=yes
-ac_cv_file__dev_zero=yes
-ac_cv_func___va_copy=no
-ac_cv_func__exit=yes
-ac_cv_func_bcopy=yes
-ac_cv_func_bzero=yes
-ac_cv_func_bcmp=yes
-ac_cv_func_creal=yes
-ac_cv_func_cimag=yes
-ac_cv_func_fchmod=yes
-ac_cv_func_getaddrinfo=yes
-ac_cv_func_getcwd=yes
-ac_cv_func_getdomainname=yes
-ac_cv_func_getpgrp_void=yes
-ac_cv_func_getpwuid_r=yes
-ac_cv_func_gettimeofday=yes
-ac_cv_func_index=yes
-ac_cv_func_lstat_dereferences_slashed_symlink=yes
-ac_cv_func_lstat_empty_string_bug=no
-ac_cv_func_lstat=yes
-ac_cv_func_malloc_0_nonnull=yes
-ac_cv_func_malloc_works=yes
-ac_cv_func_memcmp_clean=yes
-ac_cv_func_memcmp_working=yes
-ac_cv_func_posix_getgrgid_r=yes
-ac_cv_func_posix_getpwuid_r=yes
-ac_cv_func_psignal=yes
-ac_cv_func_pthread_key_delete=yes
-ac_cv_func_realloc_0_nonnull=yes
-ac_cv_func_realloc_works=yes
-ac_cv_func_rename=yes
-ac_cv_func_rindex=yes
-ac_cv_func_setlocale=yes
-ac_cv_func_setgrent_void=yes
-ac_cv_func_setpgrp_void=yes
-ac_cv_func_setresuid=no
-ac_cv_func_setvbuf_reversed=no
-ac_cv_func_stat_empty_string_bug=no
-ac_cv_func_stat_ignores_trailing_slash=no
-ac_cv_func_strerror=yes
-ac_cv_func_strftime=yes
-ac_cv_func_utimes=yes
-ac_cv_func___adjtimex=yes
-ac_cv_func_va_copy=no
-ac_cv_func_vsnprintf=yes
-ac_cv_have_accrights_in_msghdr=no
-ac_cv_have_broken_snprintf=no
-ac_cv_have_control_in_msghdr=yes
-ac_cv_have_decl_sys_siglist=no
-ac_cv_have_openpty_ctty_bug=yes
-ac_cv_have_space_d_name_in_struct_dirent=yes
-ac_cv_header_netinet_sctp_h=no
-ac_cv_header_netinet_sctp_uio_h=no
-ac_cv_int64_t=yes
-ac_cv_lbl_unaligned_fail=no
-ac_cv_linux_kernel_pppoe=yes
-ac_cv_linux_vers=2
-ac_cv_pack_bitfields_reversed=yes
-ac_cv_path_LDCONFIG=
-ac_cv_regexec_segfault_emptystr=no
-ac_cv_sctp=no
-ac_cv_sys_restartable_syscalls=yes
-ac_cv_time_r_type=POSIX
-ac_cv_type_suseconds_t=yes
-ac_cv_uchar=no
-ac_cv_uint=yes
-ac_cv_uint64_t=yes
-ac_cv_ulong=yes
-ac_cv_ushort=yes
-ac_cv_va_copy=C99
-ac_cv_va_val_copy=yes
-as_cv_unaligned_access=yes
-ac_cv_func_malloc_0_nonnull=yes
-ac_cv_func_realloc_0_nonnull=yes
diff --git a/include/site/m68k b/include/site/m68k
deleted file mode 100644
index 0037600a09..0000000000
--- a/include/site/m68k
+++ /dev/null
@@ -1,28 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/mips b/include/site/mips
deleted file mode 100644
index a5626a722b..0000000000
--- a/include/site/mips
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/mips64 b/include/site/mips64
deleted file mode 100644
index 19e75ab991..0000000000
--- a/include/site/mips64
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=8
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=8
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=8
-ac_cv_sizeof_ssize_t=8
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=8
diff --git a/include/site/mips64el b/include/site/mips64el
deleted file mode 100644
index c5aa9c5543..0000000000
--- a/include/site/mips64el
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=8
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=8
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=8
-ac_cv_sizeof_ssize_t=8
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=8
diff --git a/include/site/mipsel b/include/site/mipsel
deleted file mode 100644
index 72a3805c64..0000000000
--- a/include/site/mipsel
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/powerpc b/include/site/powerpc
deleted file mode 100644
index a5626a722b..0000000000
--- a/include/site/powerpc
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/powerpc64 b/include/site/powerpc64
deleted file mode 100644
index 6a66346b42..0000000000
--- a/include/site/powerpc64
+++ /dev/null
@@ -1,26 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof_char=1
-ac_cv_sizeof_char_p=8
-ac_cv_sizeof_double=8
-ac_cv_sizeof_float=4
-ac_cv_sizeof_int=4
-ac_cv_sizeof_long=8
-ac_cv_sizeof_long_double=16
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long_long_int=8
-ac_cv_sizeof_short=2
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_signed_char=1
-ac_cv_sizeof_unsigned_char=1
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_int=8
-ac_cv_sizeof_unsigned_long_long_int=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_unsigned_short_int=2
-ac_cv_sizeof_void_p=8
diff --git a/include/site/sparc b/include/site/sparc
deleted file mode 100644
index a5626a722b..0000000000
--- a/include/site/sparc
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=no}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=yes}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=4
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=4
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=4
-ac_cv_sizeof_ssize_t=4
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=4
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=4
diff --git a/include/site/x86_64 b/include/site/x86_64
deleted file mode 100644
index b8d581d448..0000000000
--- a/include/site/x86_64
+++ /dev/null
@@ -1,30 +0,0 @@
-#!/bin/sh
-. $TOPDIR/include/site/linux
-ac_cv_c_littleendian=${ac_cv_c_littleendian=yes}
-ac_cv_c_bigendian=${ac_cv_c_bigendian=no}
-
-ac_cv_sizeof___int64=0
-ac_cv_sizeof_char=1
-ac_cv_sizeof_int=4
-ac_cv_sizeof_int16_t=2
-ac_cv_sizeof_int32_t=4
-ac_cv_sizeof_int64_t=8
-ac_cv_sizeof_long_int=8
-ac_cv_sizeof_long_long=8
-ac_cv_sizeof_long=8
-ac_cv_sizeof_off_t=8
-ac_cv_sizeof_short_int=2
-ac_cv_sizeof_short=2
-ac_cv_sizeof_size_t=8
-ac_cv_sizeof_ssize_t=8
-ac_cv_sizeof_u_int16_t=2
-ac_cv_sizeof_u_int32_t=4
-ac_cv_sizeof_u_int64_t=8
-ac_cv_sizeof_uint16_t=2
-ac_cv_sizeof_uint32_t=4
-ac_cv_sizeof_uint64_t=8
-ac_cv_sizeof_unsigned_int=4
-ac_cv_sizeof_unsigned_long=8
-ac_cv_sizeof_unsigned_long_long=8
-ac_cv_sizeof_unsigned_short=2
-ac_cv_sizeof_void_p=8
