diff --git a/libstdc++-v3/include/Makefile.in b/libstdc++-v3/include/Makefile.in
old mode 100644
new mode 100755
index 790887a25..ecfa8ea91
--- a/libstdc++-v3/include/Makefile.in
+++ b/libstdc++-v3/include/Makefile.in
@@ -1477,7 +1477,7 @@ stamp-std: ${std_headers}
 
 stamp-bits: ${bits_headers}
 	@-mkdir -p ${bits_builddir}
-	@-cd ${bits_builddir} && $(LN_S) $? . 2>/dev/null
+	@-cd ${bits_builddir} && $(LN_S) $(filter-out stamp-bits,$?) . 2>/dev/null
 	@$(STAMP) stamp-bits
 
 stamp-bits-sup: stamp-bits ${bits_sup_headers}
@@ -1491,7 +1491,7 @@ stamp-c_base: ${c_base_headers}
 
 stamp-c_base_extra: ${c_base_headers_extra}
 	@-mkdir -p ${bits_builddir}
-	@-cd ${bits_builddir} && $(LN_S) $? . 2>/dev/null
+	@-cd ${bits_builddir} && $(LN_S) $(filter-out stamp-bits,$?) . 2>/dev/null
 	@$(STAMP) stamp-c_base_extra
 
 stamp-c_compatibility: ${c_compatibility_headers_extra}
