From 8baf583a3f2cc33988cdf54c8b931ce16bc8dcbc Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 23 Apr 2020 23:14:45 +0100
Subject: [PATCH 02/12] Allow IPv6 addresses ofthe form [::ffff:1.2.3.4] in
 --dhcp-option.

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
---
 src/option.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/option.c b/src/option.c
index 1f698da..fed17a9 100644
--- a/src/option.c
+++ b/src/option.c
@@ -1225,7 +1225,7 @@ static int parse_dhcp_opt(char *errstr, char *arg, int flags)
 	  } 
 	else if (c == '.')	
 	  {
-	    is_addr6 = is_dec = is_hex = 0;
+	    is_dec = is_hex = 0;
 	    dots++;
 	  }
 	else if (c == '-')
-- 
2.24.2 (Apple Git-127)

