diff --git a/src/Makefile.am b/src/Makefile.am
index c43d1bd..b2e60b0 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -31,7 +31,7 @@ CONFIG_FILES = \
     $()
 
 default-config.h default-config.c: bin2c $(CONFIG_FILES)
-	$(BIN2C_V)./bin2c -n lnav_config_json default-config $(CONFIG_FILES)
+	$(BIN2C_V)bin2c -n lnav_config_json default-config $(CONFIG_FILES)
 
 BUILTIN_LNAVSCRIPTS = \
     $(srcdir)/scripts/dhclient-summary.lnav \
@@ -41,29 +41,29 @@ BUILTIN_LNAVSCRIPTS = \
     $()
 
 builtin-scripts.h builtin-scripts.c: bin2c $(BUILTIN_LNAVSCRIPTS)
-	$(BIN2C_V)./bin2c -n lnav_scripts builtin-scripts $(BUILTIN_LNAVSCRIPTS)
+	$(BIN2C_V)bin2c -n lnav_scripts builtin-scripts $(BUILTIN_LNAVSCRIPTS)
 
 BUILTIN_SHSCRIPTS = \
     $(srcdir)/scripts/dump-pid.sh \
     $()
 
 builtin-sh-scripts.h builtin-sh-scripts.c: bin2c $(BUILTIN_SHSCRIPTS)
-	$(BIN2C_V)./bin2c -n lnav_sh_scripts builtin-sh-scripts $(BUILTIN_SHSCRIPTS)
+	$(BIN2C_V)bin2c -n lnav_sh_scripts builtin-sh-scripts $(BUILTIN_SHSCRIPTS)
 
 %-sh.c: $(srcdir)/%.sh bin2c
-	$(BIN2C_V)./bin2c $(*)-sh $<
+	$(BIN2C_V)bin2c $(*)-sh $<
 
 %-txt.c %-txt.h: $(srcdir)/%.txt bin2c
-	$(BIN2C_V)./bin2c $(*)-txt $<
+	$(BIN2C_V)bin2c $(*)-txt $<
 
 %-sql.c %-sql.h: $(srcdir)/%.sql bin2c
-	$(BIN2C_V)./bin2c $(*)-sql $<
+	$(BIN2C_V)bin2c $(*)-sql $<
 
 %-lnav.c %-lnav.h: $(srcdir)/%.lnav bin2c
-	$(BIN2C_V)./bin2c $(*)-lnav $<
+	$(BIN2C_V)bin2c $(*)-lnav $<
 
 %-json.c %-json.h: $(srcdir)/%.json bin2c
-	$(BIN2C_V)./bin2c $(*)-json $<
+	$(BIN2C_V)bin2c $(*)-json $<
 
 TIME_FORMATS = \
     "@%@" \
