diff --git a/configure.ac b/configure.ac
index 4302526..0f0ca0a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -96,7 +96,7 @@ dnl Internationalization macros.
 AM_GNU_GETTEXT([external], [need-ngettext])
 AM_GNU_GETTEXT_VERSION(0.18)
 
-AC_CHECK_HEADERS([sys/file.h sys/termios.h string.h locale.h getopt.h])
+AC_CHECK_HEADERS([sys/file.h string.h locale.h getopt.h])
 
 AC_CHECK_LIB(dbm, main)
 AC_CHECK_LIB(ndbm, main)
diff --git a/src/gdbmconst.h b/src/gdbmconst.h
index d30da3a..5ce182a 100644
--- a/src/gdbmconst.h
+++ b/src/gdbmconst.h
@@ -54,5 +54,7 @@
 /* The size of the bucket cache. */
 #define DEFAULT_CACHESIZE  100
 
+#ifndef SIZE_T_MAX
 /* Maximum size representable by a size_t variable */
-#define SIZE_T_MAX ((size_t)-1)
+# define SIZE_T_MAX ((size_t)-1)
+#endif
diff --git a/src/gdbmtool.c b/src/gdbmtool.c
index 9bd23ea..ceffaef 100644
--- a/src/gdbmtool.c
+++ b/src/gdbmtool.c
@@ -24,9 +24,7 @@
 #include <signal.h>
 #include <pwd.h>
 #include <sys/ioctl.h>
-#ifdef HAVE_SYS_TERMIOS_H
-# include <sys/termios.h>
-#endif
+#include <termios.h>
 #include <stdarg.h>
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
@@ -38,8 +36,6 @@ datum key_data;               /* Current key */
 datum return_data;            /* Current data */
 int open_mode;                /* Default open mode */
 
-#define SIZE_T_MAX ((size_t)-1)
-
 unsigned input_line;
 
 
diff --git a/src/mmap.c b/src/mmap.c
index 3d57b11..273f64c 100644
--- a/src/mmap.c
+++ b/src/mmap.c
@@ -33,6 +33,14 @@
 #  define MAP_FAILED ((void*)-1)
 # endif
 
+# if defined(MAP_POPULATE)
+#  define GDBM_MMAP_FLAGS MAP_POPULATE
+# elif defined(MAP_PREFAULT_READ)
+#  define GDBM_MMAP_FLAGS MAP_PREFAULT_READ
+# else
+#  define GDBM_MMAP_FLAGS 0
+# endif
+
 /* Translate current offset in the mapped region into the absolute position */
 # define _GDBM_MMAPPED_POS(dbf) ((dbf)->mapped_off + (dbf)->mapped_pos)
 /* Return true if the absolute offset OFF lies within the currentlty mmapped
@@ -90,7 +98,7 @@ int
 _gdbm_internal_remap (GDBM_FILE dbf, size_t size)
 {
   void *p;
-  int flags = PROT_READ;
+  int prot = PROT_READ;
   size_t page_size = sysconf (_SC_PAGESIZE);
 
   if (dbf->mapped_region)
@@ -107,9 +115,9 @@ _gdbm_internal_remap (GDBM_FILE dbf, size_t size)
   dbf->mapped_off = (dbf->mapped_off / page_size) * page_size;
 
   if (dbf->read_write)
-    flags |= PROT_WRITE;
+    prot |= PROT_WRITE;
   
-  p = mmap (NULL, dbf->mapped_size, flags, MAP_SHARED,
+  p = mmap (NULL, dbf->mapped_size, prot, MAP_SHARED | GDBM_MMAP_FLAGS,
 	    dbf->desc, dbf->mapped_off);
   if (p == MAP_FAILED)
     {
diff --git a/tests/closerr.c b/tests/closerr.c
index 39fb0b7..b86d74f 100644
--- a/tests/closerr.c
+++ b/tests/closerr.c
@@ -21,6 +21,7 @@
 #include <string.h>
 #include <errno.h>
 #include <assert.h>
+#include <unistd.h>
 #include "gdbm.h"
 
 int
@@ -28,7 +29,6 @@ main (int argc, char **argv)
 {
   GDBM_FILE dbf;
   char dbname[] = "junk.gdbm";
-  int rc;
   
   assert (argc == 1);
   
