diff --git a/Makefile.def b/Makefile.def
index 36fd26b03..4825df3a5 100644
--- a/Makefile.def
+++ b/Makefile.def
@@ -137,7 +137,8 @@ host_modules= { module= libctf; no_install=true; no_check=true;
 target_modules = { module= libstdc++-v3;
 		   bootstrap=true;
 		   lib_path=src/.libs;
-		   raw_cxx=true; };
+		   raw_cxx=true;
+		   bootstrap_configure_flags='--disable-libstdcxx-debug --disable-libstdcxx-pch'; };
 target_modules = { module= libsanitizer;
 		   bootstrap=true;
 		   lib_path=.libs;
diff --git a/Makefile.tpl b/Makefile.tpl
index efed15117..1c570f038 100644
--- a/Makefile.tpl
+++ b/Makefile.tpl
@@ -1145,7 +1145,9 @@ configure-stage[+id+]-[+prefix+][+module+]:
 	  --target=[+target_alias+] \
 	  [+ IF prev +]--with-build-libsubdir=$(HOST_SUBDIR)[+ ENDIF prev +] \
 	  $(STAGE[+id+]_CONFIGURE_FLAGS)[+ IF extra_configure_flags +] \
-	  [+extra_configure_flags+][+ ENDIF extra_configure_flags +]
+	  [+extra_configure_flags+][+ ENDIF extra_configure_flags +] \
+	  [+ IF bootstrap_configure_flags +][+bootstrap_configure_flags+] \
+	  [+ ENDIF bootstrap_configure_flags +]
 @endif [+prefix+][+module+]-bootstrap
 [+ ENDFOR bootstrap_stage +]
 [+ ENDIF bootstrap +]
