diff --git a/Makefile.in b/Makefile.in
index 64e091ba7..b4de280e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -372,7 +372,7 @@ BUILD_PREFIX_1 = @BUILD_PREFIX_1@
 
 # Flags to pass to stage2 and later makes.  They are defined
 # here so that they can be overridden by Makefile fragments.
-BOOT_CFLAGS= -g -O2
+BOOT_CFLAGS= -g1 -O2
 BOOT_LDFLAGS=
 BOOT_ADAFLAGS= -gnatpg
 
diff --git a/Makefile.tpl b/Makefile.tpl
index 41cae58a2..ebab7a9b3 100644
--- a/Makefile.tpl
+++ b/Makefile.tpl
@@ -375,7 +375,7 @@ BUILD_PREFIX_1 = @BUILD_PREFIX_1@
 
 # Flags to pass to stage2 and later makes.  They are defined
 # here so that they can be overridden by Makefile fragments.
-BOOT_CFLAGS= -g -O2
+BOOT_CFLAGS= -g1 -O2
 BOOT_LDFLAGS=
 BOOT_ADAFLAGS= -gnatpg
 
diff --git a/config/acinclude.m4 b/config/acinclude.m4
index 8242b2c7a..42553c3b5 100644
--- a/config/acinclude.m4
+++ b/config/acinclude.m4
@@ -24,7 +24,7 @@ dnl normal versions of a library), tasteless as that idea is.
   if test "$ac_test_CXXFLAGS" = set; then
     CXXFLAGS="$ac_save_CXXFLAGS"
   elif test $ac_cv_prog_cxx_g = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-O2"
   fi
@@ -121,7 +121,7 @@ dnl normal versions of a library), tasteless as that idea is.
   if test "$ac_test_CFLAGS" = set; then
     CFLAGS="$ac_save_CFLAGS"
   elif test $ac_cv_prog_cc_g = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-O2"
   fi
diff --git a/configure b/configure
index abd93a990..c9999e214 100755
--- a/configure
+++ b/configure
@@ -4711,7 +4711,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5058,7 +5058,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
@@ -6830,7 +6830,7 @@ fi
 # CFLAGS_FOR_TARGET and CXXFLAGS_FOR_TARGET.
 if test "x$CFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CFLAGS_FOR_TARGET="-g -O2"
+    CFLAGS_FOR_TARGET="-g1 -O2"
   else
     CFLAGS_FOR_TARGET=$CFLAGS
     case " $CFLAGS " in
@@ -6847,7 +6847,7 @@ fi
 
 if test "x$CXXFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CXXFLAGS_FOR_TARGET="-g -O2"
+    CXXFLAGS_FOR_TARGET="-g1 -O2"
   else
     CXXFLAGS_FOR_TARGET=$CXXFLAGS
     case " $CXXFLAGS " in
diff --git a/configure.ac b/configure.ac
index 9db4fd14a..52496e5ce 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2351,7 +2351,7 @@ AC_SUBST(DEBUG_PREFIX_CFLAGS_FOR_TARGET)
 # CFLAGS_FOR_TARGET and CXXFLAGS_FOR_TARGET.
 if test "x$CFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CFLAGS_FOR_TARGET="-g -O2"
+    CFLAGS_FOR_TARGET="-g1 -O2"
   else
     CFLAGS_FOR_TARGET=$CFLAGS
     case " $CFLAGS " in
@@ -2368,7 +2368,7 @@ AC_SUBST(CFLAGS_FOR_TARGET)
 
 if test "x$CXXFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CXXFLAGS_FOR_TARGET="-g -O2"
+    CXXFLAGS_FOR_TARGET="-g1 -O2"
   else
     CXXFLAGS_FOR_TARGET=$CXXFLAGS
     case " $CXXFLAGS " in
diff --git a/fixincludes/configure b/fixincludes/configure
index 6e2d67b65..c007662a3 100755
--- a/fixincludes/configure
+++ b/fixincludes/configure
@@ -3987,7 +3987,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/gcc/ada/gcc-interface/Make-lang.in b/gcc/ada/gcc-interface/Make-lang.in
index de23b1410..48f3307b6 100644
--- a/gcc/ada/gcc-interface/Make-lang.in
+++ b/gcc/ada/gcc-interface/Make-lang.in
@@ -70,7 +70,7 @@ FORCE_DEBUG_ADAFLAGS = -g
 ADA_CFLAGS =
 ADA_INCLUDES = -nostdinc -I- -I. -Iada/generated -Iada -I$(srcdir)/ada -I$(srcdir)/ada/gcc-interface -Iada/libgnat -I$(srcdir)/ada/libgnat
 GNATLIBFLAGS= -W -Wall -gnatpg -nostdinc
-GNATLIBCFLAGS= -g -O2 $(TCFLAGS)
+GNATLIBCFLAGS= -g1 -O2 $(TCFLAGS)
 ADA_INCLUDE_DIR = $(libsubdir)/adainclude
 ADA_RTL_OBJ_DIR = $(libsubdir)/adalib
 THREAD_KIND=native
diff --git a/gcc/ada/gcc-interface/Makefile.in b/gcc/ada/gcc-interface/Makefile.in
index a0a5bb241..4991536d5 100644
--- a/gcc/ada/gcc-interface/Makefile.in
+++ b/gcc/ada/gcc-interface/Makefile.in
@@ -111,7 +111,7 @@ NO_OMIT_ADAFLAGS = -fno-omit-frame-pointer
 NO_SIBLING_ADAFLAGS = -fno-optimize-sibling-calls
 NO_REORDER_ADAFLAGS = -fno-toplevel-reorder
 GNATLIBFLAGS = -W -Wall -gnatpg -nostdinc
-GNATLIBCFLAGS = -g -O2
+GNATLIBCFLAGS = -g1 -O2
 # Pretend that _Unwind_GetIPInfo is available for the target by default.  This
 # should be autodetected during the configuration of libada and passed down to
 # here, but we need something for --disable-libada and hope for the best.
diff --git a/gcc/ada/libgnat/g-comlin.ads b/gcc/ada/libgnat/g-comlin.ads
index 8306f9a00..8ba890b0b 100644
--- a/gcc/ada/libgnat/g-comlin.ads
+++ b/gcc/ada/libgnat/g-comlin.ads
@@ -543,7 +543,7 @@ package GNAT.Command_Line is
    -----------------
 
    --  The following subprograms are used to manipulate a command line
-   --  represented as a string (for instance "-g -O2"), as well as parsing
+   --  represented as a string (for instance "-g1 -O2"), as well as parsing
    --  the switches from such a string. They provide high-level configurations
    --  to define aliases (a switch is equivalent to one or more other switches)
    --  or grouping of switches ("-gnatyac" is equivalent to "-gnatya" and
diff --git a/gcc/configure b/gcc/configure
index a065ba237..895f12eec 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -4527,7 +4527,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4874,7 +4874,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/gcc/doc/install.texi b/gcc/doc/install.texi
index 0d08696c2..4af1b5801 100644
--- a/gcc/doc/install.texi
+++ b/gcc/doc/install.texi
@@ -2482,7 +2482,7 @@ make BOOT_CFLAGS='-O' bootstrap
 @end smallexample
 
 You can place non-default optimization flags into @code{BOOT_CFLAGS}; they
-are less well tested here than the default of @samp{-g -O2}, but should
+are less well tested here than the default of @samp{-g1 -O2}, but should
 still work.  In a few cases, you may find that you need to specify special
 flags such as @option{-msoft-float} here to complete the bootstrap; or,
 if the native compiler miscompiles the stage1 compiler, you may need
diff --git a/gcc/testsuite/g++.dg/cpp0x/pr83734.C b/gcc/testsuite/g++.dg/cpp0x/pr83734.C
index 958c2e962..906e0052c 100644
--- a/gcc/testsuite/g++.dg/cpp0x/pr83734.C
+++ b/gcc/testsuite/g++.dg/cpp0x/pr83734.C
@@ -1,6 +1,6 @@
 // PR c++/83734
 // { dg-do compile { target c++11 } }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 struct A { constexpr A () { typedef int T; } };
 A a;
diff --git a/gcc/testsuite/g++.dg/ipa/pr65034.C b/gcc/testsuite/g++.dg/ipa/pr65034.C
index cb33d5056..c07c58013 100644
--- a/gcc/testsuite/g++.dg/ipa/pr65034.C
+++ b/gcc/testsuite/g++.dg/ipa/pr65034.C
@@ -1,6 +1,6 @@
 // PR ipa/65034
 // { dg-do compile }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 enum B { C };
 enum D { E };
diff --git a/gcc/testsuite/g++.dg/lto/pr48508-1_0.C b/gcc/testsuite/g++.dg/lto/pr48508-1_0.C
index 6faee75b1..01ffe8234 100644
--- a/gcc/testsuite/g++.dg/lto/pr48508-1_0.C
+++ b/gcc/testsuite/g++.dg/lto/pr48508-1_0.C
@@ -1,5 +1,5 @@
 // { dg-lto-do link }
-// { dg-lto-options { { -g -O2 -flto -flto-partition=none } } }
+// { dg-lto-options { { -g1 -O2 -flto -flto-partition=none } } }
 
 void __attribute__((externally_visible))
 foo (int i)
diff --git a/gcc/testsuite/g++.dg/lto/pr65549_0.C b/gcc/testsuite/g++.dg/lto/pr65549_0.C
index 382f7eff7..44936ccc5 100644
--- a/gcc/testsuite/g++.dg/lto/pr65549_0.C
+++ b/gcc/testsuite/g++.dg/lto/pr65549_0.C
@@ -1,6 +1,6 @@
 // { dg-lto-do link }
 // { dg-require-effective-target lto_incremental }
-// { dg-lto-options { { -std=gnu++14 -flto -g -Wno-return-type } { -std=gnu++14 -flto -g -O2 -fno-inline -flto-partition=max -Wno-return-type } } }
+// { dg-lto-options { { -std=gnu++14 -flto -g -Wno-return-type } { -std=gnu++14 -flto -g1 -O2 -fno-inline -flto-partition=max -Wno-return-type } } }
 // { dg-extra-ld-options "-r -nostdlib" }
 
 namespace std {
diff --git a/gcc/testsuite/g++.dg/opt/pr48967.C b/gcc/testsuite/g++.dg/opt/pr48967.C
index db2ea5474..84cc66bc6 100644
--- a/gcc/testsuite/g++.dg/opt/pr48967.C
+++ b/gcc/testsuite/g++.dg/opt/pr48967.C
@@ -1,6 +1,6 @@
 // PR debug/48967
 // { dg-do compile }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 template <typename> struct A;
 template <typename T> struct A <T *>
diff --git a/gcc/testsuite/g++.dg/other/crash-8.C b/gcc/testsuite/g++.dg/other/crash-8.C
index b7f56fc9b..7b42c053d 100644
--- a/gcc/testsuite/g++.dg/other/crash-8.C
+++ b/gcc/testsuite/g++.dg/other/crash-8.C
@@ -1,6 +1,6 @@
 // Origin: PR c++/42797
 // { dg-do compile { target c++11 } }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 template<typename _Tp, _Tp __v>     struct integral_constant     {
     static const _Tp value = __v;
diff --git a/gcc/testsuite/g++.dg/other/pr43631.C b/gcc/testsuite/g++.dg/other/pr43631.C
index eb4f57861..c2a1e3579 100644
--- a/gcc/testsuite/g++.dg/other/pr43631.C
+++ b/gcc/testsuite/g++.dg/other/pr43631.C
@@ -1,6 +1,6 @@
 // PR middle-end/43631
 // { dg-do compile }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 // { dg-additional-options "-mtune=atom" { target i?86-*-* x86_64-*-* } }
 
 typedef void (*T) ();
diff --git a/gcc/testsuite/g++.dg/pr61033.C b/gcc/testsuite/g++.dg/pr61033.C
index 6974351a9..9d9a8f6b5 100644
--- a/gcc/testsuite/g++.dg/pr61033.C
+++ b/gcc/testsuite/g++.dg/pr61033.C
@@ -1,6 +1,6 @@
 // PR debug/61033
 // { dg-do compile }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 # 0 "" 3
 struct A {
diff --git a/gcc/testsuite/g++.dg/pr82155.C b/gcc/testsuite/g++.dg/pr82155.C
index 75d9b615f..120c0f465 100644
--- a/gcc/testsuite/g++.dg/pr82155.C
+++ b/gcc/testsuite/g++.dg/pr82155.C
@@ -1,5 +1,5 @@
 /* { dg-do compile { target c++11 } } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 template <typename a> struct b { a c; };
 template <typename d> struct e { d *operator->(); };
diff --git a/gcc/testsuite/g++.old-deja/g++.other/crash3.C b/gcc/testsuite/g++.old-deja/g++.other/crash3.C
index 5a2182716..efaae1b00 100644
--- a/gcc/testsuite/g++.old-deja/g++.other/crash3.C
+++ b/gcc/testsuite/g++.old-deja/g++.other/crash3.C
@@ -1,5 +1,5 @@
 // { dg-do assemble  }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 inline void f() {
   struct S {};
diff --git a/gcc/testsuite/g++.old-deja/g++.other/debug6.C b/gcc/testsuite/g++.old-deja/g++.other/debug6.C
index a4d088413..152dda636 100644
--- a/gcc/testsuite/g++.old-deja/g++.other/debug6.C
+++ b/gcc/testsuite/g++.old-deja/g++.other/debug6.C
@@ -1,5 +1,5 @@
 // { dg-do assemble  }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 
 //  Copyright (C) 1999 Free Software Foundation, Inc.
 //  Contributed by Nathan Sidwell 21 Nov 1999 <nathan@acm.org>
diff --git a/gcc/testsuite/g++.old-deja/g++.other/debug7.C b/gcc/testsuite/g++.old-deja/g++.other/debug7.C
index f486de1ab..eb2a43168 100644
--- a/gcc/testsuite/g++.old-deja/g++.other/debug7.C
+++ b/gcc/testsuite/g++.old-deja/g++.other/debug7.C
@@ -1,5 +1,5 @@
 // { dg-do link  }
-// { dg-options "-g -O2" }
+// { dg-options "-g1 -O2" }
 // Origin: Loren James Rittle <rittle@latour.rsch.comm.mot.com>
 
 namespace std {
diff --git a/gcc/testsuite/gcc.dg/debug/dwarf2/pr88635.c b/gcc/testsuite/gcc.dg/debug/dwarf2/pr88635.c
index 02c53d100..0b1267dee 100644
--- a/gcc/testsuite/gcc.dg/debug/dwarf2/pr88635.c
+++ b/gcc/testsuite/gcc.dg/debug/dwarf2/pr88635.c
@@ -1,6 +1,6 @@
 /* PR debug/88635 */
 /* { dg-do assemble } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 /* { dg-additional-options "-fpie" { target pie } } */
 
 static void
diff --git a/gcc/testsuite/gcc.dg/delay-slot-2.c b/gcc/testsuite/gcc.dg/delay-slot-2.c
index d8619aac9..25aabcbeb 100644
--- a/gcc/testsuite/gcc.dg/delay-slot-2.c
+++ b/gcc/testsuite/gcc.dg/delay-slot-2.c
@@ -2,7 +2,7 @@
 /* Reported by Jurij Smakov <jurij@wooyd.org> */
 
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 extern int printf (__const char *__restrict __format, ...);
 extern void print_c_condition (const char *);
diff --git a/gcc/testsuite/gcc.dg/dfp/pr48928.c b/gcc/testsuite/gcc.dg/dfp/pr48928.c
index 1fd363565..17077e804 100644
--- a/gcc/testsuite/gcc.dg/dfp/pr48928.c
+++ b/gcc/testsuite/gcc.dg/dfp/pr48928.c
@@ -1,6 +1,6 @@
 /* PR debug/48928 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 _Decimal32
 foo (_Decimal32 x)
diff --git a/gcc/testsuite/gcc.dg/graphite/pr41888.c b/gcc/testsuite/gcc.dg/graphite/pr41888.c
index 45df7df28..8e55c9feb 100644
--- a/gcc/testsuite/gcc.dg/graphite/pr41888.c
+++ b/gcc/testsuite/gcc.dg/graphite/pr41888.c
@@ -1,4 +1,4 @@
-/* { dg-options "-g -O2 -ftree-loop-distribution -fgraphite-identity" } */
+/* { dg-options "-g1 -O2 -ftree-loop-distribution -fgraphite-identity" } */
 
 int
 foo (int *x)
diff --git a/gcc/testsuite/gcc.dg/pr42662.c b/gcc/testsuite/gcc.dg/pr42662.c
index 6f416e826..96f60c5c1 100644
--- a/gcc/testsuite/gcc.dg/pr42662.c
+++ b/gcc/testsuite/gcc.dg/pr42662.c
@@ -1,6 +1,6 @@
 /* PR debug/42662 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 struct S { unsigned long s[17]; };
 
diff --git a/gcc/testsuite/gcc.dg/pr42715.c b/gcc/testsuite/gcc.dg/pr42715.c
index 72bdfc862..5ed686ed5 100644
--- a/gcc/testsuite/gcc.dg/pr42715.c
+++ b/gcc/testsuite/gcc.dg/pr42715.c
@@ -1,5 +1,5 @@
 /* { dg-do compile { target fpic } } */
-/* { dg-options "-fPIC -g -O2 -w" } */
+/* { dg-options "-fPIC -g1 -O2 -w" } */
 /* var-tracking failed to clobber the reg holding v at the asm insn,
    so v ended up bound to an intermediate PIC expression.  */
 
diff --git a/gcc/testsuite/gcc.dg/pr43058.c b/gcc/testsuite/gcc.dg/pr43058.c
index bba6f814a..295c6649b 100644
--- a/gcc/testsuite/gcc.dg/pr43058.c
+++ b/gcc/testsuite/gcc.dg/pr43058.c
@@ -1,6 +1,6 @@
 /* PR debug/43058 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 /* { dg-timeout-factor 4 } */
 
 extern void *f1 (void *, void *, void *);
diff --git a/gcc/testsuite/gcc.dg/pr43139.c b/gcc/testsuite/gcc.dg/pr43139.c
index 4dc725bcc..9a5275e37 100644
--- a/gcc/testsuite/gcc.dg/pr43139.c
+++ b/gcc/testsuite/gcc.dg/pr43139.c
@@ -1,6 +1,6 @@
 /* PR target/43139 */
 /* { dg-do compile { target fpic } } */
-/* { dg-options "-g -O2 -fpic" } */
+/* { dg-options "-g1 -O2 -fpic" } */
 
 typedef double T1[10];
 typedef double T2[10][10];
diff --git a/gcc/testsuite/gcc.dg/pr43299.c b/gcc/testsuite/gcc.dg/pr43299.c
index d71c89689..00b1a4c96 100644
--- a/gcc/testsuite/gcc.dg/pr43299.c
+++ b/gcc/testsuite/gcc.dg/pr43299.c
@@ -1,6 +1,6 @@
 /* PR debug/43299 */
 /* { dg-do assemble } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 extern void *emit_insn (void *);
 
diff --git a/gcc/testsuite/gcc.dg/pr43557-1.c b/gcc/testsuite/gcc.dg/pr43557-1.c
index 5ba467d01..417480fd2 100644
--- a/gcc/testsuite/gcc.dg/pr43557-1.c
+++ b/gcc/testsuite/gcc.dg/pr43557-1.c
@@ -1,7 +1,7 @@
 /* PR debug/43557 */
 /* { dg-do link } */
 /* { dg-require-effective-target lto } */
-/* { dg-options "-flto -g -O2" } */
+/* { dg-options "-flto -g1 -O2" } */
 /* { dg-additional-sources "pr43557-2.c" } */
 
 struct S
diff --git a/gcc/testsuite/gcc.dg/pr45259.c b/gcc/testsuite/gcc.dg/pr45259.c
index 7a051af81..51204ac82 100644
--- a/gcc/testsuite/gcc.dg/pr45259.c
+++ b/gcc/testsuite/gcc.dg/pr45259.c
@@ -1,7 +1,7 @@
 /* PR debug/45259 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -w -fpic" { target fpic } } */
-/* { dg-options "-g -O2 -w" { target { ! fpic } } } */
+/* { dg-options "-g1 -O2 -w -fpic" { target fpic } } */
+/* { dg-options "-g1 -O2 -w" { target { ! fpic } } } */
 /* { dg-require-effective-target indirect_jumps } */
 /* { dg-require-effective-target label_values } */
 
diff --git a/gcc/testsuite/gcc.dg/pr46387.c b/gcc/testsuite/gcc.dg/pr46387.c
index e06bd3b10..e52824d89 100644
--- a/gcc/testsuite/gcc.dg/pr46387.c
+++ b/gcc/testsuite/gcc.dg/pr46387.c
@@ -1,6 +1,6 @@
 /* PR debug/46387 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 struct S { double x; double y; short z; };
 int a = 0, b = 0, c;
diff --git a/gcc/testsuite/gcc.dg/pr48213.c b/gcc/testsuite/gcc.dg/pr48213.c
index cf7d76fa0..99744a589 100644
--- a/gcc/testsuite/gcc.dg/pr48213.c
+++ b/gcc/testsuite/gcc.dg/pr48213.c
@@ -1,7 +1,7 @@
 /* PR target/48213 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
-/* { dg-options "-g -O2 -fpic" { target fpic } } */
+/* { dg-options "-g1 -O2" } */
+/* { dg-options "-g1 -O2 -fpic" { target fpic } } */
 
 struct S { int (*s) (void); };
 int bar (void);
diff --git a/gcc/testsuite/gcc.dg/pr49544.c b/gcc/testsuite/gcc.dg/pr49544.c
index 52f79dd7f..bda806ec5 100644
--- a/gcc/testsuite/gcc.dg/pr49544.c
+++ b/gcc/testsuite/gcc.dg/pr49544.c
@@ -1,6 +1,6 @@
 /* PR debug/49544 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 /* { dg-require-effective-target ptr32plus } */
 
 __extension__ typedef __PTRDIFF_TYPE__ ptr_t;
diff --git a/gcc/testsuite/gcc.dg/pr49602.c b/gcc/testsuite/gcc.dg/pr49602.c
index bb8eeb0c3..7ccd058ad 100644
--- a/gcc/testsuite/gcc.dg/pr49602.c
+++ b/gcc/testsuite/gcc.dg/pr49602.c
@@ -1,6 +1,6 @@
 /* PR debug/49602 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 static void
 foo (int *x)
diff --git a/gcc/testsuite/gcc.dg/pr50827.c b/gcc/testsuite/gcc.dg/pr50827.c
index 5d258bfcf..4782d7ded 100644
--- a/gcc/testsuite/gcc.dg/pr50827.c
+++ b/gcc/testsuite/gcc.dg/pr50827.c
@@ -1,6 +1,6 @@
 /* PR debug/50827 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -funroll-loops" } */
+/* { dg-options "-g1 -O2 -funroll-loops" } */
 
 void
 foo (int w, int x, int *y, int *z)
diff --git a/gcc/testsuite/gcc.dg/pr63342.c b/gcc/testsuite/gcc.dg/pr63342.c
index 499d508c1..080411290 100644
--- a/gcc/testsuite/gcc.dg/pr63342.c
+++ b/gcc/testsuite/gcc.dg/pr63342.c
@@ -1,6 +1,6 @@
 /* PR debug/63342 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 /* { dg-additional-options "-fpic" { target fpic } } */
 
 static __thread double u[9], v[9];
diff --git a/gcc/testsuite/gcc.dg/pr80463.c b/gcc/testsuite/gcc.dg/pr80463.c
index cebf2fef1..c8acd7d1c 100644
--- a/gcc/testsuite/gcc.dg/pr80463.c
+++ b/gcc/testsuite/gcc.dg/pr80463.c
@@ -1,5 +1,5 @@
 /* { dg-do compile { target powerpc*-*-* ia64-*-* i?86-*-* x86_64-*-* } } */
-/* { dg-options "-g -O2 -fvar-tracking-assignments -fselective-scheduling2 -ftree-loop-vectorize -fnon-call-exceptions -fno-tree-vrp -fno-gcse-lm -fno-tree-loop-im -fno-reorder-blocks-and-partition -fno-reorder-blocks -fno-move-loop-invariants -w" } */
+/* { dg-options "-g1 -O2 -fvar-tracking-assignments -fselective-scheduling2 -ftree-loop-vectorize -fnon-call-exceptions -fno-tree-vrp -fno-gcse-lm -fno-tree-loop-im -fno-reorder-blocks-and-partition -fno-reorder-blocks -fno-move-loop-invariants -w" } */
 
 short int t2;
 int cd, aa, ft;
diff --git a/gcc/testsuite/gcc.dg/pr83723.c b/gcc/testsuite/gcc.dg/pr83723.c
index a64fe9b56..fd5bd855f 100644
--- a/gcc/testsuite/gcc.dg/pr83723.c
+++ b/gcc/testsuite/gcc.dg/pr83723.c
@@ -1,6 +1,6 @@
 /* PR rtl-optimization/83723 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 /* { dg-additional-options "-mfpmath=sse -msse2" { target i?86-*-* x86_64-*-* } } */
 /* { dg-additional-options "-fpie" { target pie } } */
 
diff --git a/gcc/testsuite/gcc.dg/pr83986.c b/gcc/testsuite/gcc.dg/pr83986.c
index 31a53d936..7eec9f67f 100644
--- a/gcc/testsuite/gcc.dg/pr83986.c
+++ b/gcc/testsuite/gcc.dg/pr83986.c
@@ -1,6 +1,6 @@
 /* PR rtl-optimization/83986 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -fsched2-use-superblocks -funwind-tables --param max-pending-list-length=1" } */
+/* { dg-options "-g1 -O2 -fsched2-use-superblocks -funwind-tables --param max-pending-list-length=1" } */
 
 int v;
 
diff --git a/gcc/testsuite/gcc.dg/pr84300.c b/gcc/testsuite/gcc.dg/pr84300.c
index 6016799a4..35e439433 100644
--- a/gcc/testsuite/gcc.dg/pr84300.c
+++ b/gcc/testsuite/gcc.dg/pr84300.c
@@ -1,5 +1,5 @@
 /* { dg-do compile } */
 /* { dg-require-effective-target split_stack } */
-/* { dg-options "-g -O2 -fsplit-stack -fno-omit-frame-pointer" } */
+/* { dg-options "-g1 -O2 -fsplit-stack -fno-omit-frame-pointer" } */
 
 void trap () { __builtin_trap (); }
diff --git a/gcc/testsuite/gcc.dg/pr84721.c b/gcc/testsuite/gcc.dg/pr84721.c
index 3ebf3a6c8..a55e4cf1d 100644
--- a/gcc/testsuite/gcc.dg/pr84721.c
+++ b/gcc/testsuite/gcc.dg/pr84721.c
@@ -1,6 +1,6 @@
 /* PR c/84721 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 int a[({ int b })];		/* { dg-error "braced-group within expression allowed only inside a function" } */
 int c[({ int d () {}; })];	/* { dg-error "braced-group within expression allowed only inside a function" } */
diff --git a/gcc/testsuite/gcc.dg/pr86064.c b/gcc/testsuite/gcc.dg/pr86064.c
index 339742725..b7a6884c5 100644
--- a/gcc/testsuite/gcc.dg/pr86064.c
+++ b/gcc/testsuite/gcc.dg/pr86064.c
@@ -1,6 +1,6 @@
 /* { dg-do compile } */
 /* { dg-skip-if "split DWARF unsupported" { *-*-darwin* } } */
-/* { dg-options "-g -O2 -fno-var-tracking-assignments -gsplit-dwarf" } */
+/* { dg-options "-g1 -O2 -fno-var-tracking-assignments -gsplit-dwarf" } */
 
 /* This used to fail with location views (implicitly) enabled, because
    var-tracking (not at assignments) creates a range for d starting at
diff --git a/gcc/testsuite/gcc.dg/pr86459.c b/gcc/testsuite/gcc.dg/pr86459.c
index 7856a372d..200c924a2 100644
--- a/gcc/testsuite/gcc.dg/pr86459.c
+++ b/gcc/testsuite/gcc.dg/pr86459.c
@@ -1,5 +1,5 @@
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -fno-var-tracking-assignments -gsplit-dwarf -g3" } */
+/* { dg-options "-g1 -O2 -fno-var-tracking-assignments -gsplit-dwarf -g3" } */
 
 /* Same as pr86064.c but compiled with -g3 it showed an issue in
    output_macinfo_op because of a typo in an assert.  */
diff --git a/gcc/testsuite/gcc.dg/pr90733.c b/gcc/testsuite/gcc.dg/pr90733.c
index 7d2a7fb55..d44c8d5b7 100644
--- a/gcc/testsuite/gcc.dg/pr90733.c
+++ b/gcc/testsuite/gcc.dg/pr90733.c
@@ -1,6 +1,6 @@
 /* PR debug/90733 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -w" } */
+/* { dg-options "-g1 -O2 -w" } */
 
 struct S { unsigned a : 1; };
 union U { struct S b; _Complex unsigned c; };
diff --git a/gcc/testsuite/gcc.target/aarch64/pr84252.c b/gcc/testsuite/gcc.target/aarch64/pr84252.c
index fc66a269d..80c4e200c 100644
--- a/gcc/testsuite/gcc.target/aarch64/pr84252.c
+++ b/gcc/testsuite/gcc.target/aarch64/pr84252.c
@@ -1,6 +1,6 @@
 /* PR debug/84252 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2" } */
+/* { dg-options "-g1 -O2" } */
 
 struct S { __Int32x4_t b[2]; };
 
diff --git a/gcc/testsuite/gcc.target/i386/pr49567.c b/gcc/testsuite/gcc.target/i386/pr49567.c
index 309deb479..1e3dcfedb 100644
--- a/gcc/testsuite/gcc.target/i386/pr49567.c
+++ b/gcc/testsuite/gcc.target/i386/pr49567.c
@@ -1,6 +1,6 @@
 /* PR debug/49567 */
 /* { dg-do compile } */
-/* { dg-options "-g -O2 -msse4" } */
+/* { dg-options "-g1 -O2 -msse4" } */
 
 #include <x86intrin.h>
 
diff --git a/gcc/testsuite/gcc.target/powerpc/pr83862.c b/gcc/testsuite/gcc.target/powerpc/pr83862.c
index 3cadb5796..edb1faec8 100644
--- a/gcc/testsuite/gcc.target/powerpc/pr83862.c
+++ b/gcc/testsuite/gcc.target/powerpc/pr83862.c
@@ -8,7 +8,7 @@
    memory could abort if the memory address was indexed (reg+reg).  The
    optimization is only on 64-bit machines with direct move.
 
-   Compile with -g -O2 -mabi=ieeelongdouble -Wno-psabi.  */
+   Compile with -g1 -O2 -mabi=ieeelongdouble -Wno-psabi.  */
 
 #ifndef TYPE
 #define TYPE long double
diff --git a/gcc/testsuite/gdc.test/gdc-test.exp b/gcc/testsuite/gdc.test/gdc-test.exp
index ce5575c07..053876f1b 100644
--- a/gcc/testsuite/gdc.test/gdc-test.exp
+++ b/gcc/testsuite/gdc.test/gdc-test.exp
@@ -362,7 +362,7 @@ proc gdc-do-test { } {
     # If a testcase doesn't have special options, use these.
     global DEFAULT_DFLAGS
     if ![info exists DEFAULT_DFLAGS] then {
-	set DEFAULT_DFLAGS "-g -O2 -frelease"
+	set DEFAULT_DFLAGS "-g1 -O2 -frelease"
 	#set DEFAULT_DFLAGS "-O2"
     }
 
diff --git a/gcc/testsuite/gfortran.dg/pr57393-1.f90 b/gcc/testsuite/gfortran.dg/pr57393-1.f90
index 6b2cb1b3f..65430d461 100644
--- a/gcc/testsuite/gfortran.dg/pr57393-1.f90
+++ b/gcc/testsuite/gfortran.dg/pr57393-1.f90
@@ -1,6 +1,6 @@
 ! PR middle-end/57393
 ! { dg-do compile }
-! { dg-options "-g -O2 -ffast-math" }
+! { dg-options "-g1 -O2 -ffast-math" }
 
 SUBROUTINE pr57393(nn,e,g,t0,t1,t2,t3,t4,t5,t6,t7,&
                    t8,t9,t10,t11,t12,t13,t14,t15,&
diff --git a/gcc/testsuite/gfortran.dg/pr57393-2.f90 b/gcc/testsuite/gfortran.dg/pr57393-2.f90
index fafa8f900..4192892e3 100644
--- a/gcc/testsuite/gfortran.dg/pr57393-2.f90
+++ b/gcc/testsuite/gfortran.dg/pr57393-2.f90
@@ -1,6 +1,6 @@
 ! PR middle-end/57393
 ! { dg-do compile }
-! { dg-options "-g -O2" }
+! { dg-options "-g1 -O2" }
 
 SUBROUTINE pr57393 ( a1, a2, a3, a4, a5, a6, a7 )
   COMPLEX(kind=8), DIMENSION(:), INTENT(IN) :: a1
diff --git a/gnattools/configure b/gnattools/configure
index c2d755b72..504583a01 100755
--- a/gnattools/configure
+++ b/gnattools/configure
@@ -2809,7 +2809,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/gotools/configure b/gotools/configure
index ff71d2ae8..5c214737a 100755
--- a/gotools/configure
+++ b/gotools/configure
@@ -3384,7 +3384,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -3958,7 +3958,7 @@ $as_echo "$ac_try_echo"; } >&5
   rm -f conftest.er1 conftest.err
   $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
   test $ac_status = 0; }
-GOFLAGS="-g -O2"
+GOFLAGS="-g1 -O2"
 ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
 ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
diff --git a/intl/configure b/intl/configure
index 2f3599314..cc29f4b35 100755
--- a/intl/configure
+++ b/intl/configure
@@ -2888,7 +2888,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libada/Makefile.in b/libada/Makefile.in
index 522b92073..98e7f02c3 100644
--- a/libada/Makefile.in
+++ b/libada/Makefile.in
@@ -59,7 +59,7 @@ LDFLAGS=
 CFLAGS=-g
 PICFLAG = @PICFLAG@
 GNATLIBFLAGS= -W -Wall -gnatpg -nostdinc
-GNATLIBCFLAGS= -g -O2
+GNATLIBCFLAGS= -g1 -O2
 GNATLIBCFLAGS_FOR_C = -W -Wall $(GNATLIBCFLAGS) $(CFLAGS_FOR_TARGET) \
 	-fexceptions -DIN_RTS @have_getipinfo@ @have_capability@
 
diff --git a/libada/configure b/libada/configure
index 743f98afd..2114dc574 100755
--- a/libada/configure
+++ b/libada/configure
@@ -2987,7 +2987,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libatomic/configure b/libatomic/configure
index e7076a0ff..1339054d9 100755
--- a/libatomic/configure
+++ b/libatomic/configure
@@ -4063,7 +4063,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libbacktrace/configure b/libbacktrace/configure
index 51d7cb458..871497645 100755
--- a/libbacktrace/configure
+++ b/libbacktrace/configure
@@ -3389,7 +3389,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5063,7 +5063,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libcc1/configure b/libcc1/configure
index 53f40d311..cdf5a9efc 100755
--- a/libcc1/configure
+++ b/libcc1/configure
@@ -3240,7 +3240,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -11359,7 +11359,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libcpp/configure b/libcpp/configure
index a7aa35c5b..e4d38a349 100755
--- a/libcpp/configure
+++ b/libcpp/configure
@@ -3550,7 +3550,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -3897,7 +3897,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libdecnumber/configure b/libdecnumber/configure
index 14c103a07..41596894b 100755
--- a/libdecnumber/configure
+++ b/libdecnumber/configure
@@ -2999,7 +2999,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libffi/configure b/libffi/configure
index 2324890a5..229f553a8 100755
--- a/libffi/configure
+++ b/libffi/configure
@@ -4197,7 +4197,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4794,7 +4794,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libgcc/configure b/libgcc/configure
index 8ceec1d04..7b754b3c2 100755
--- a/libgcc/configure
+++ b/libgcc/configure
@@ -3909,7 +3909,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libgfortran/configure b/libgfortran/configure
index 60867b93d..3ce727088 100755
--- a/libgfortran/configure
+++ b/libgfortran/configure
@@ -4431,7 +4431,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5664,7 +5664,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -13260,7 +13260,7 @@ if test "$ac_test_FCFLAGS" = set; then
   FCFLAGS=$ac_save_FCFLAGS
 elif test $ac_cv_prog_fc_g = yes; then
   if test "x$ac_cv_fc_compiler_gnu" = xyes; then
-    FCFLAGS="-g -O2"
+    FCFLAGS="-g1 -O2"
   else
     FCFLAGS="-g"
   fi
diff --git a/libgo/config/libtool.m4 b/libgo/config/libtool.m4
index e74475373..74310ae05 100644
--- a/libgo/config/libtool.m4
+++ b/libgo/config/libtool.m4
@@ -7126,7 +7126,7 @@ AC_DEFUN([LT_PROG_GCJ],
 [m4_ifdef([AC_PROG_GCJ], [AC_PROG_GCJ],
   [m4_ifdef([A][M_PROG_GCJ], [A][M_PROG_GCJ],
     [AC_CHECK_TOOL(GCJ, gcj,)
-      test "x${GCJFLAGS+set}" = xset || GCJFLAGS="-g -O2"
+      test "x${GCJFLAGS+set}" = xset || GCJFLAGS="-g1 -O2"
       AC_SUBST(GCJFLAGS)])])[]dnl
 ])
 
diff --git a/libgo/configure b/libgo/configure
index 0d5845e15..3fe7261d5 100755
--- a/libgo/configure
+++ b/libgo/configure
@@ -3975,7 +3975,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4549,7 +4549,7 @@ $as_echo "$ac_try_echo"; } >&5
   rm -f conftest.er1 conftest.err
   $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
   test $ac_status = 0; }
-GOFLAGS="-g -O2"
+GOFLAGS="-g1 -O2"
 ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
 ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
diff --git a/libgo/go/cmd/go/internal/work/exec.go b/libgo/go/cmd/go/internal/work/exec.go
index 530272466..1da05bb68 100644
--- a/libgo/go/cmd/go/internal/work/exec.go
+++ b/libgo/go/cmd/go/internal/work/exec.go
@@ -2393,7 +2393,7 @@ func envList(key, def string) []string {
 
 // CFlags returns the flags to use when invoking the C, C++ or Fortran compilers, or cgo.
 func (b *Builder) CFlags(p *load.Package) (cppflags, cflags, cxxflags, fflags, ldflags []string, err error) {
-	defaults := "-g -O2"
+	defaults := "-g1 -O2"
 
 	if cppflags, err = buildFlags("CPPFLAGS", "", p.CgoCPPFLAGS, checkCompilerFlags); err != nil {
 		return
diff --git a/libgo/go/debug/dwarf/testdata/line-gcc-win.bin b/libgo/go/debug/dwarf/testdata/line-gcc-win.bin
index 583ad44dd..1fe08b0db 100644
Binary files a/libgo/go/debug/dwarf/testdata/line-gcc-win.bin and b/libgo/go/debug/dwarf/testdata/line-gcc-win.bin differ
diff --git a/libgo/go/debug/dwarf/testdata/ranges.c b/libgo/go/debug/dwarf/testdata/ranges.c
index 2f208e591..46c955c0b 100644
--- a/libgo/go/debug/dwarf/testdata/ranges.c
+++ b/libgo/go/debug/dwarf/testdata/ranges.c
@@ -1,4 +1,4 @@
-// gcc -g -O2 -freorder-blocks-and-partition
+// gcc -g1 -O2 -freorder-blocks-and-partition
 
 const char *arr[10000];
 const char *hot = "hot";
diff --git a/libgo/go/debug/dwarf/testdata/ranges.elf b/libgo/go/debug/dwarf/testdata/ranges.elf
index 7f54138cf..485d5aac6 100644
Binary files a/libgo/go/debug/dwarf/testdata/ranges.elf and b/libgo/go/debug/dwarf/testdata/ranges.elf differ
diff --git a/libgomp/configure b/libgomp/configure
index b4bc4f436..1b921572f 100755
--- a/libgomp/configure
+++ b/libgomp/configure
@@ -4214,7 +4214,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -11991,7 +11991,7 @@ if test "$ac_test_FCFLAGS" = set; then
   FCFLAGS=$ac_save_FCFLAGS
 elif test $ac_cv_prog_fc_g = yes; then
   if test "x$ac_cv_fc_compiler_gnu" = xyes; then
-    FCFLAGS="-g -O2"
+    FCFLAGS="-g1 -O2"
   else
     FCFLAGS="-g"
   fi
diff --git a/libhsail-rt/configure b/libhsail-rt/configure
index 6a086102d..abf9b5ef4 100755
--- a/libhsail-rt/configure
+++ b/libhsail-rt/configure
@@ -3884,7 +3884,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4481,7 +4481,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libiberty/configure b/libiberty/configure
index 7a34dabec..37ba2f020 100755
--- a/libiberty/configure
+++ b/libiberty/configure
@@ -3691,7 +3691,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libitm/configure b/libitm/configure
index fb742d79e..b1f074a20 100644
--- a/libitm/configure
+++ b/libitm/configure
@@ -4247,7 +4247,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4844,7 +4844,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libobjc/configure b/libobjc/configure
index c122c817d..7c7102711 100755
--- a/libobjc/configure
+++ b/libobjc/configure
@@ -3271,7 +3271,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/liboffloadmic/configure b/liboffloadmic/configure
index 52b551779..325107300 100644
--- a/liboffloadmic/configure
+++ b/liboffloadmic/configure
@@ -3842,7 +3842,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4439,7 +4439,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/liboffloadmic/plugin/configure b/liboffloadmic/plugin/configure
index 2ae61621c..2dc7f42fa 100644
--- a/liboffloadmic/plugin/configure
+++ b/liboffloadmic/plugin/configure
@@ -3700,7 +3700,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4297,7 +4297,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libphobos/configure b/libphobos/configure
index e2977b3cc..585709fba 100755
--- a/libphobos/configure
+++ b/libphobos/configure
@@ -3308,7 +3308,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4957,7 +4957,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libquadmath/configure b/libquadmath/configure
index 5283adbdd..e3681f464 100755
--- a/libquadmath/configure
+++ b/libquadmath/configure
@@ -3838,7 +3838,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -11642,7 +11642,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libsanitizer/configure b/libsanitizer/configure
index 226da62cf..d676c1f77 100755
--- a/libsanitizer/configure
+++ b/libsanitizer/configure
@@ -3451,7 +3451,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5475,7 +5475,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5881,7 +5881,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libssp/configure b/libssp/configure
index 2adbcc3fc..a3b7f491d 100755
--- a/libssp/configure
+++ b/libssp/configure
@@ -3779,7 +3779,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/libstdc++-v3/ChangeLog-1999 b/libstdc++-v3/ChangeLog-1999
index 846535bf9..8efd842cf 100644
--- a/libstdc++-v3/ChangeLog-1999
+++ b/libstdc++-v3/ChangeLog-1999
@@ -3185,7 +3185,7 @@
 1999-02-22  Benjamin Kosnik  <bkoz@loony.cygnus.com>
 
 	* mkcshadow: And change copyright here.
-	* src/Makefile.am (CXXFLAGS): Add -g -O2. Add new header
+	* src/Makefile.am (CXXFLAGS): Add -g1 -O2. Add new header
 	dependencies for stl_pthread_alloc.h, pthread_allocimpl.h,
 	stl_thread.h, and stl_iterator_base.h.
 	* src/Makefile.in (CXXFLAGS): Ditto.
diff --git a/libstdc++-v3/ChangeLog-2003 b/libstdc++-v3/ChangeLog-2003
index 3def08eef..7f61f7fed 100644
--- a/libstdc++-v3/ChangeLog-2003
+++ b/libstdc++-v3/ChangeLog-2003
@@ -6946,8 +6946,8 @@
 	* include/Makefile.am (stamp-std-precompile): Generate stdc++.h.gch
 	* include/Makefile.in: Regenerate.
 	* testsuite_flags.in (--build-cxx): Add, but don't use PCHFLAGS.
-	(--cxxflags): Put -g -O2 here.
-	* testsuite/libstdc++-v3.dg/dg.exp: Remove -g -O2 here.
+	(--cxxflags): Put -g1 -O2 here.
+	* testsuite/libstdc++-v3.dg/dg.exp: Remove -g1 -O2 here.
 
 2003-03-14  Loren J. Rittle  <ljrittle@acm.org>
 
@@ -8425,7 +8425,7 @@
 	* testsuite/lib/libstdc++-v3-dg.exp
 	(libstdc++-v3-list-sourcefiles): Remove wchar_t files if
 	testsuite_wchar_t is not present in the build directory.
-	* testsuite/libstdc++-v3.dg/dg.exp: Add -g -O2 to DEFAULT_CXXFLAGS.
+	* testsuite/libstdc++-v3.dg/dg.exp: Add -g1 -O2 to DEFAULT_CXXFLAGS.
 	Remove setulimit bits.
 	* testsuite/26_numerics/complex_value.cc: Set to noopts.
 	* testsuite/Makefile.am (CLEANFILES): Add.
diff --git a/libstdc++-v3/ChangeLog-2012 b/libstdc++-v3/ChangeLog-2012
index 3b53aece3..6809baeaf 100644
--- a/libstdc++-v3/ChangeLog-2012
+++ b/libstdc++-v3/ChangeLog-2012
@@ -3199,7 +3199,7 @@
 
 2012-03-23  Benjamin Kosnik  <bkoz@redhat.com>
 
-	* scripts/testsuite_flags.in (--cxxflags): Let CXXFLAGS set -g -O2.
+	* scripts/testsuite_flags.in (--cxxflags): Let CXXFLAGS set -g1 -O2.
 	* testsuite/lib/libstdc++.exp: Same.
 
 2012-03-23  Paolo Carlini  <paolo.carlini@oracle.com>
diff --git a/libstdc++-v3/configure b/libstdc++-v3/configure
index 1225edc59..f87d854cb 100755
--- a/libstdc++-v3/configure
+++ b/libstdc++-v3/configure
@@ -4505,7 +4505,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4911,7 +4911,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/libstdc++-v3/doc/html/manual/abi.html b/libstdc++-v3/doc/html/manual/abi.html
index 96383c331..c456796f3 100644
--- a/libstdc++-v3/doc/html/manual/abi.html
+++ b/libstdc++-v3/doc/html/manual/abi.html
@@ -496,7 +496,7 @@ exceptions, locale, etc.
   Then, the "C" compiler is used to compile a source file that uses
   functions from each library.
 </p><pre class="programlisting">
-gcc test.c -g -O2 -L. -lone -ltwo /usr/lib/libstdc++.so.5 /usr/lib/libstdc++.so.6
+gcc test.c -g1 -O2 -L. -lone -ltwo /usr/lib/libstdc++.so.5 /usr/lib/libstdc++.so.6
 </pre><p>
   Which gives the expected:
 </p><pre class="screen">
diff --git a/libstdc++-v3/doc/html/manual/debug.html b/libstdc++-v3/doc/html/manual/debug.html
index bce4242cd..d71c82288 100644
--- a/libstdc++-v3/doc/html/manual/debug.html
+++ b/libstdc++-v3/doc/html/manual/debug.html
@@ -8,7 +8,7 @@
     between compilation and debug or analysis tools.
   </p><p>
     The default optimizations and debug flags for a libstdc++ build
-    are <code class="code">-g -O2</code>. However, both debug and optimization
+    are <code class="code">-g1 -O2</code>. However, both debug and optimization
     flags can be varied to change debugging characteristics. For
     instance, turning off all optimization via the <code class="code">-g -O0
     -fno-inline</code> flags will disable inlining and optimizations,
diff --git a/libstdc++-v3/doc/html/manual/using_headers.html b/libstdc++-v3/doc/html/manual/using_headers.html
index 9cc1c19bb..d40daa83d 100644
--- a/libstdc++-v3/doc/html/manual/using_headers.html
+++ b/libstdc++-v3/doc/html/manual/using_headers.html
@@ -151,12 +151,12 @@ g++ -v hello.cc
 End of search list.
 </pre><p>Then, create a precompiled header file with the same flags that
 will be used to compile other projects.</p><pre class="programlisting">
-g++ -Winvalid-pch -x c++-header -g -O2 -o ./stdc++.h.gch /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/x86_64-unknown-linux-gnu/bits/stdc++.h
+g++ -Winvalid-pch -x c++-header -g1 -O2 -o ./stdc++.h.gch /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/x86_64-unknown-linux-gnu/bits/stdc++.h
 </pre><p>The resulting file will be quite large: the current size is around
 thirty megabytes. </p><p>How to use the resulting file.</p><pre class="programlisting">
-g++ -I. -include stdc++.h  -H -g -O2 hello.cc
+g++ -I. -include stdc++.h  -H -g1 -O2 hello.cc
 </pre><p>Verification that the PCH file is being used is easy:</p><pre class="programlisting">
-g++ -Winvalid-pch -I. -include stdc++.h -H -g -O2 hello.cc -o test.exe
+g++ -Winvalid-pch -I. -include stdc++.h -H -g1 -O2 hello.cc -o test.exe
 ! ./stdc++.h.gch
 . /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/iostream
 . /mnt/share/bld/H-x86-gcc.20071201include/c++/4.3.0/string
diff --git a/libstdc++-v3/doc/xml/manual/abi.xml b/libstdc++-v3/doc/xml/manual/abi.xml
index 969edd7c8..193982b6f 100644
--- a/libstdc++-v3/doc/xml/manual/abi.xml
+++ b/libstdc++-v3/doc/xml/manual/abi.xml
@@ -1061,7 +1061,7 @@ exceptions, locale, etc.
   functions from each library.
 </para>
 <programlisting>
-gcc test.c -g -O2 -L. -lone -ltwo /usr/lib/libstdc++.so.5 /usr/lib/libstdc++.so.6
+gcc test.c -g1 -O2 -L. -lone -ltwo /usr/lib/libstdc++.so.5 /usr/lib/libstdc++.so.6
 </programlisting>
 
 <para>
diff --git a/libstdc++-v3/doc/xml/manual/debug.xml b/libstdc++-v3/doc/xml/manual/debug.xml
index 091e0b691..c3746a056 100644
--- a/libstdc++-v3/doc/xml/manual/debug.xml
+++ b/libstdc++-v3/doc/xml/manual/debug.xml
@@ -26,7 +26,7 @@
 
   <para>
     The default optimizations and debug flags for a libstdc++ build
-    are <code>-g -O2</code>. However, both debug and optimization
+    are <code>-g1 -O2</code>. However, both debug and optimization
     flags can be varied to change debugging characteristics. For
     instance, turning off all optimization via the <code>-g -O0
     -fno-inline</code> flags will disable inlining and optimizations,
diff --git a/libstdc++-v3/doc/xml/manual/using.xml b/libstdc++-v3/doc/xml/manual/using.xml
index 8e099d9c3..6b3e9588f 100644
--- a/libstdc++-v3/doc/xml/manual/using.xml
+++ b/libstdc++-v3/doc/xml/manual/using.xml
@@ -971,7 +971,7 @@ End of search list.
 will be used to compile other projects.</para>
 
 <programlisting>
-g++ -Winvalid-pch -x c++-header -g -O2 -o ./stdc++.h.gch /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/x86_64-unknown-linux-gnu/bits/stdc++.h
+g++ -Winvalid-pch -x c++-header -g1 -O2 -o ./stdc++.h.gch /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/x86_64-unknown-linux-gnu/bits/stdc++.h
 </programlisting>
 
 <para>The resulting file will be quite large: the current size is around
@@ -980,13 +980,13 @@ thirty megabytes. </para>
 <para>How to use the resulting file.</para>
 
 <programlisting>
-g++ -I. -include stdc++.h  -H -g -O2 hello.cc
+g++ -I. -include stdc++.h  -H -g1 -O2 hello.cc
 </programlisting>
 
 <para>Verification that the PCH file is being used is easy:</para>
 
 <programlisting>
-g++ -Winvalid-pch -I. -include stdc++.h -H -g -O2 hello.cc -o test.exe
+g++ -Winvalid-pch -I. -include stdc++.h -H -g1 -O2 hello.cc -o test.exe
 ! ./stdc++.h.gch
 . /mnt/share/bld/H-x86-gcc.20071201/include/c++/4.3.0/iostream
 . /mnt/share/bld/H-x86-gcc.20071201include/c++/4.3.0/string
diff --git a/libstdc++-v3/testsuite/18_support/bad_exception/23591_thread-1.c b/libstdc++-v3/testsuite/18_support/bad_exception/23591_thread-1.c
index a6f40871d..4419b23bb 100644
--- a/libstdc++-v3/testsuite/18_support/bad_exception/23591_thread-1.c
+++ b/libstdc++-v3/testsuite/18_support/bad_exception/23591_thread-1.c
@@ -1,6 +1,6 @@
 // { dg-require-sharedlib "" }
 // { dg-skip-if "invalid options for C" { *-*-* } { "-std=c++??" "-std=gnu++??" } }
-// { dg-options "-g -O2 -pthread -ldl -x c -fvtable-verify=none -Wno-pedantic" { target *-*-linux* *-*-gnu* *-*-solaris2.1[2-9]* } }
+// { dg-options "-g1 -O2 -pthread -ldl -x c -fvtable-verify=none -Wno-pedantic" { target *-*-linux* *-*-gnu* *-*-solaris2.1[2-9]* } }
 
 // Copyright (C) 2005-2019 Free Software Foundation, Inc.
 //
diff --git a/libstdc++-v3/testsuite/ext/mt_allocator/22309_thread.cc b/libstdc++-v3/testsuite/ext/mt_allocator/22309_thread.cc
index 7ca1f2173..993177e60 100644
--- a/libstdc++-v3/testsuite/ext/mt_allocator/22309_thread.cc
+++ b/libstdc++-v3/testsuite/ext/mt_allocator/22309_thread.cc
@@ -1,5 +1,5 @@
 // { dg-require-sharedlib "" }
-// { dg-options "-g -O2 -pthread -ldl" { target *-*-linux* *-*-gnu* } }
+// { dg-options "-g1 -O2 -pthread -ldl" { target *-*-linux* *-*-gnu* } }
 
 // Copyright (C) 2004-2019 Free Software Foundation, Inc.
 //
diff --git a/libstdc++-v3/testsuite/lib/libstdc++.exp b/libstdc++-v3/testsuite/lib/libstdc++.exp
index a616227ae..e514a32b8 100644
--- a/libstdc++-v3/testsuite/lib/libstdc++.exp
+++ b/libstdc++-v3/testsuite/lib/libstdc++.exp
@@ -309,7 +309,7 @@ proc libstdc++_init { testfile } {
     append cxxflags [getenv CXXFLAGS]
 
     if ![regexp ".*-O" $cxxflags] {
-	append cxxflags " -g -O2"
+	append cxxflags " -g1 -O2"
     }
 
     v3track cxxflags 2
diff --git a/libtool.m4 b/libtool.m4
index 896676288..61d251e91 100644
--- a/libtool.m4
+++ b/libtool.m4
@@ -7052,7 +7052,7 @@ AC_DEFUN([LT_PROG_GCJ],
 [m4_ifdef([AC_PROG_GCJ], [AC_PROG_GCJ],
   [m4_ifdef([A][M_PROG_GCJ], [A][M_PROG_GCJ],
     [AC_CHECK_TOOL(GCJ, gcj,)
-      test "x${GCJFLAGS+set}" = xset || GCJFLAGS="-g -O2"
+      test "x${GCJFLAGS+set}" = xset || GCJFLAGS="-g1 -O2"
       AC_SUBST(GCJFLAGS)])])[]dnl
 ])
 
diff --git a/libvtv/configure b/libvtv/configure
index bfdf34954..396abcce8 100755
--- a/libvtv/configure
+++ b/libvtv/configure
@@ -3292,7 +3292,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5311,7 +5311,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5717,7 +5717,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/lto-plugin/configure b/lto-plugin/configure
index 8fdcab96d..ec941cee9 100755
--- a/lto-plugin/configure
+++ b/lto-plugin/configure
@@ -3897,7 +3897,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5082,7 +5082,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/zlib/configure b/zlib/configure
index 7a8b49b0b..c8b20cc4c 100755
--- a/zlib/configure
+++ b/zlib/configure
@@ -3777,7 +3777,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
