diff -Naur miniupnpd-2.1-orig/Makefile.linux miniupnpd-2.1/Makefile.linux
--- miniupnpd-2.1-orig/Makefile.linux	2020-06-28 09:45:17.747420125 +0300
+++ miniupnpd-2.1/Makefile.linux	2020-06-28 09:46:59.958383048 +0300
@@ -158,8 +158,6 @@
 
 # OpenWrt packager disables https server for IGD v2 and hardcodes libuuid support
 ifeq ($(TARGET_OPENWRT),)
-LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l libssl)
-
 TEST := $(shell $(PKG_CONFIG) --exists uuid && echo 1)
 ifeq ($(TEST),1)
 LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l uuid)
