--- a/tools/lib/bpf/btf.c
+++ b/tools/lib/bpf/btf.c
@@ -7,6 +7,7 @@
 #include <string.h>
 #include <fcntl.h>
 #include <unistd.h>
+#include <asm/errno.h>
 #include <errno.h>
 #include <sys/utsname.h>
 #include <sys/param.h>
--- a/tools/lib/bpf/btf_dump.c
+++ b/tools/lib/bpf/btf_dump.c
@@ -10,6 +10,7 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#include <asm/errno.h>
 #include <errno.h>
 #include <linux/err.h>
 #include <linux/btf.h>
--- a/tools/lib/bpf/hashmap.c
+++ b/tools/lib/bpf/hashmap.c
@@ -8,6 +8,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <stdio.h>
+#include <asm/errno.h>
 #include <errno.h>
 #include <linux/err.h>
 #include "hashmap.h"
--- a/tools/lib/bpf/libbpf.c
+++ b/tools/lib/bpf/libbpf.c
@@ -23,6 +23,7 @@
 #include <unistd.h>
 #include <endian.h>
 #include <fcntl.h>
+#include <asm/errno.h>
 #include <errno.h>
 #include <ctype.h>
 #include <asm/unistd.h>
--- a/tools/lib/bpf/ringbuf.c
+++ b/tools/lib/bpf/ringbuf.c
@@ -9,6 +9,7 @@
 #endif
 #include <stdlib.h>
 #include <stdio.h>
+#include <asm/errno.h>
 #include <errno.h>
 #include <unistd.h>
 #include <linux/err.h>
