diff --git a/Makefile.in b/Makefile.in
index 64e091ba7..b4de280e5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -372,7 +372,7 @@ BUILD_PREFIX_1 = @BUILD_PREFIX_1@
 
 # Flags to pass to stage2 and later makes.  They are defined
 # here so that they can be overridden by Makefile fragments.
-BOOT_CFLAGS= -g -O2
+BOOT_CFLAGS= -g1 -O2
 BOOT_LDFLAGS=
 BOOT_ADAFLAGS= -gnatpg
 
diff --git a/Makefile.tpl b/Makefile.tpl
index 41cae58a2..ebab7a9b3 100644
--- a/Makefile.tpl
+++ b/Makefile.tpl
@@ -375,7 +375,7 @@ BUILD_PREFIX_1 = @BUILD_PREFIX_1@
 
 # Flags to pass to stage2 and later makes.  They are defined
 # here so that they can be overridden by Makefile fragments.
-BOOT_CFLAGS= -g -O2
+BOOT_CFLAGS= -g1 -O2
 BOOT_LDFLAGS=
 BOOT_ADAFLAGS= -gnatpg
 
diff --git a/config/acinclude.m4 b/config/acinclude.m4
index 8242b2c7a..42553c3b5 100644
--- a/config/acinclude.m4
+++ b/config/acinclude.m4
@@ -24,7 +24,7 @@ dnl normal versions of a library), tasteless as that idea is.
   if test "$ac_test_CXXFLAGS" = set; then
     CXXFLAGS="$ac_save_CXXFLAGS"
   elif test $ac_cv_prog_cxx_g = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-O2"
   fi
@@ -121,7 +121,7 @@ dnl normal versions of a library), tasteless as that idea is.
   if test "$ac_test_CFLAGS" = set; then
     CFLAGS="$ac_save_CFLAGS"
   elif test $ac_cv_prog_cc_g = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-O2"
   fi
diff --git a/configure b/configure
index abd93a990..c9999e214 100755
--- a/configure
+++ b/configure
@@ -4711,7 +4711,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -5058,7 +5058,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
@@ -6830,7 +6830,7 @@ fi
 # CFLAGS_FOR_TARGET and CXXFLAGS_FOR_TARGET.
 if test "x$CFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CFLAGS_FOR_TARGET="-g -O2"
+    CFLAGS_FOR_TARGET="-g1 -O2"
   else
     CFLAGS_FOR_TARGET=$CFLAGS
     case " $CFLAGS " in
@@ -6847,7 +6847,7 @@ fi
 
 if test "x$CXXFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CXXFLAGS_FOR_TARGET="-g -O2"
+    CXXFLAGS_FOR_TARGET="-g1 -O2"
   else
     CXXFLAGS_FOR_TARGET=$CXXFLAGS
     case " $CXXFLAGS " in
diff --git a/configure.ac b/configure.ac
index 9db4fd14a..52496e5ce 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2351,7 +2351,7 @@ AC_SUBST(DEBUG_PREFIX_CFLAGS_FOR_TARGET)
 # CFLAGS_FOR_TARGET and CXXFLAGS_FOR_TARGET.
 if test "x$CFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CFLAGS_FOR_TARGET="-g -O2"
+    CFLAGS_FOR_TARGET="-g1 -O2"
   else
     CFLAGS_FOR_TARGET=$CFLAGS
     case " $CFLAGS " in
@@ -2368,7 +2368,7 @@ AC_SUBST(CFLAGS_FOR_TARGET)
 
 if test "x$CXXFLAGS_FOR_TARGET" = x; then
   if test "x${is_cross_compiler}" = xyes; then
-    CXXFLAGS_FOR_TARGET="-g -O2"
+    CXXFLAGS_FOR_TARGET="-g1 -O2"
   else
     CXXFLAGS_FOR_TARGET=$CXXFLAGS
     case " $CXXFLAGS " in
diff --git a/fixincludes/configure b/fixincludes/configure
index 6e2d67b65..c007662a3 100755
--- a/fixincludes/configure
+++ b/fixincludes/configure
@@ -3987,7 +3987,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
diff --git a/gcc/ada/gcc-interface/Make-lang.in b/gcc/ada/gcc-interface/Make-lang.in
index de23b1410..48f3307b6 100644
--- a/gcc/ada/gcc-interface/Make-lang.in
+++ b/gcc/ada/gcc-interface/Make-lang.in
@@ -70,7 +70,7 @@ FORCE_DEBUG_ADAFLAGS = -g
 ADA_CFLAGS =
 ADA_INCLUDES = -nostdinc -I- -I. -Iada/generated -Iada -I$(srcdir)/ada -I$(srcdir)/ada/gcc-interface -Iada/libgnat -I$(srcdir)/ada/libgnat
 GNATLIBFLAGS= -W -Wall -gnatpg -nostdinc
-GNATLIBCFLAGS= -g -O2 $(TCFLAGS)
+GNATLIBCFLAGS= -g1 -O2 $(TCFLAGS)
 ADA_INCLUDE_DIR = $(libsubdir)/adainclude
 ADA_RTL_OBJ_DIR = $(libsubdir)/adalib
 THREAD_KIND=native
diff --git a/gcc/ada/gcc-interface/Makefile.in b/gcc/ada/gcc-interface/Makefile.in
index a0a5bb241..4991536d5 100644
--- a/gcc/ada/gcc-interface/Makefile.in
+++ b/gcc/ada/gcc-interface/Makefile.in
@@ -111,7 +111,7 @@ NO_OMIT_ADAFLAGS = -fno-omit-frame-pointer
 NO_SIBLING_ADAFLAGS = -fno-optimize-sibling-calls
 NO_REORDER_ADAFLAGS = -fno-toplevel-reorder
 GNATLIBFLAGS = -W -Wall -gnatpg -nostdinc
-GNATLIBCFLAGS = -g -O2
+GNATLIBCFLAGS = -g1 -O2
 # Pretend that _Unwind_GetIPInfo is available for the target by default.  This
 # should be autodetected during the configuration of libada and passed down to
 # here, but we need something for --disable-libada and hope for the best.
diff --git a/gcc/ada/libgnat/g-comlin.ads b/gcc/ada/libgnat/g-comlin.ads
index 8306f9a00..8ba890b0b 100644
--- a/gcc/ada/libgnat/g-comlin.ads
+++ b/gcc/ada/libgnat/g-comlin.ads
@@ -543,7 +543,7 @@ package GNAT.Command_Line is
    -----------------
 
    --  The following subprograms are used to manipulate a command line
-   --  represented as a string (for instance "-g -O2"), as well as parsing
+   --  represented as a string (for instance "-g1 -O2"), as well as parsing
    --  the switches from such a string. They provide high-level configurations
    --  to define aliases (a switch is equivalent to one or more other switches)
    --  or grouping of switches ("-gnatyac" is equivalent to "-gnatya" and
diff --git a/gcc/configure b/gcc/configure
index a065ba237..895f12eec 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -4527,7 +4527,7 @@ if test "$ac_test_CFLAGS" = set; then
   CFLAGS=$ac_save_CFLAGS
 elif test $ac_cv_prog_cc_g = yes; then
   if test "$GCC" = yes; then
-    CFLAGS="-g -O2"
+    CFLAGS="-g1 -O2"
   else
     CFLAGS="-g"
   fi
@@ -4874,7 +4874,7 @@ if test "$ac_test_CXXFLAGS" = set; then
   CXXFLAGS=$ac_save_CXXFLAGS
 elif test $ac_cv_prog_cxx_g = yes; then
   if test "$GXX" = yes; then
-    CXXFLAGS="-g -O2"
+    CXXFLAGS="-g1 -O2"
   else
     CXXFLAGS="-g"
   fi
diff --git a/gcc/doc/install.texi b/gcc/doc/install.texi
index 0d08696c2..4af1b5801 100644
--- a/gcc/doc/install.texi
+++ b/gcc/doc/install.texi
@@ -2482,7 +2482,7 @@ make BOOT_CFLAGS='-O' bootstrap
 @end smallexample
 
 You can place non-default optimization flags into @code{BOOT_CFLAGS}; they
-are less well tested here than the default of @samp{-g -O2}, but should
+are less well tested here than the default of @samp{-g1 -O2}, but should
 still work.  In a few cases, you may find that you need to specify special
 flags such as @option{-msoft-float} here to complete the bootstrap; or,
 if the native compiler miscompiles the stage1 compiler, you may need
