diff -Naur lzo-2.10-orig/CMakeLists.txt lzo-2.10/CMakeLists.txt
--- lzo-2.10-orig/CMakeLists.txt	2018-05-11 20:47:32.230661438 +0300
+++ lzo-2.10/CMakeLists.txt	2017-03-01 22:54:14.000000000 +0300
@@ -36,14 +36,6 @@
 # // init
 # ************************************************************************/
 
-# Disallow in-source builds. Note that you will still have to manually
-# clean up a few files if you accidentally try an in-source build.
-set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
-set(CMAKE_DISABLE_SOURCE_CHANGES  ON)
-if(",${CMAKE_SOURCE_DIR}," STREQUAL ",${CMAKE_BINARY_DIR},")
-    message(FATAL_ERROR "ERROR: In-source builds are not allowed.")
-endif()
-
 if(NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE "Release" CACHE STRING "" FORCE)
 endif()
