From 7627aac4e5381546baeb0d6bef6675e9107cd751 Mon Sep 17 00:00:00 2001
From: Daniel Santos <daniel.santos@pobox.com>
Date: Sat, 25 Apr 2020 12:18:15 -0500
Subject: Fix signed compare warning

Signed-off-by: Daniel Santos <daniel.santos@pobox.com>
---
 src/thread/__timedwait.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/thread/__timedwait.c b/src/thread/__timedwait.c
index 666093be..9829b93e 100644
--- a/src/thread/__timedwait.c
+++ b/src/thread/__timedwait.c
@@ -38,7 +38,7 @@ int __timedwait_cp(volatile int *addr, int val,
 	if (priv) priv = FUTEX_PRIVATE;
 
 	if (at) {
-		if (at->tv_nsec >= 1000000000UL) return EINVAL;
+		if ((unsigned long)at->tv_nsec >= 1000000000UL) return EINVAL;
 		if (__clock_gettime(clk, &to)) return EINVAL;
 		to.tv_sec = at->tv_sec - to.tv_sec;
 		if ((to.tv_nsec = at->tv_nsec - to.tv_nsec) < 0) {
-- 
2.24.1

