--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -884,6 +884,28 @@ sub vms_info {
         ranlib           => "true",
     },
 
+# OpenWrt
+    "linux-armv4-openwrt" => {
+        inherit_from     => [ "linux-armv4" ],
+        cflags           => "-DTERMIOS \$(OPENWRT_OPTIMIZATION_FLAGS) -fomit-frame-pointer -Wall",
+    },
+    "linux-aarch64-openwrt" => {
+        inherit_from     => [ "linux-aarch64" ],
+        cflags           => "-DTERMIOS \$(OPENWRT_OPTIMIZATION_FLAGS) -fomit-frame-pointer -Wall",
+    },
+    "linux-x86_64-openwrt" => {
+        inherit_from     => [ "linux-x86_64" ],
+        cflags           => "-m64 -DL_ENDIAN -DTERMIOS \$(OPENWRT_OPTIMIZATION_FLAGS) -fomit-frame-pointer -Wall",
+    },
+    "linux-mips-openwrt" => {
+        inherit_from     => [ "linux-mips32" ],
+        cflags           => "-DTERMIOS \$(OPENWRT_OPTIMIZATION_FLAGS) -fomit-frame-pointer -Wall",
+    },
+    "linux-generic-openwrt" => {
+        inherit_from     => [ "linux-generic32" ],
+        cflags           => "-DTERMIOS \$(OPENWRT_OPTIMIZATION_FLAGS) -fomit-frame-pointer -Wall",
+    },
+
 #### Android: linux-* but without pointers to headers and libs.
     #
     # It takes pair of prior-set environment variables to make it work:
