diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1f77662..f0a9eb2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,7 +2,7 @@ cmake_minimum_required(VERSION 2.6)
 
 PROJECT(procd C)
 INCLUDE(GNUInstallDirs)
-ADD_DEFINITIONS(-Os -ggdb -Wall -Werror --std=gnu99 -Wmissing-declarations)
+ADD_DEFINITIONS(-O2 -ggdb -Wall -Werror --std=gnu99 -Wmissing-declarations)
 
 SET(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS "")
 
diff --git a/upgraded/CMakeLists.txt b/upgraded/CMakeLists.txt
index 09cf472..be50424 100644
--- a/upgraded/CMakeLists.txt
+++ b/upgraded/CMakeLists.txt
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 2.6)
 PROJECT(upgraded C)
 FIND_PATH(ubox_include_dir libubox/uloop.h)
 INCLUDE_DIRECTORIES(${ubox_include_dir})
-ADD_DEFINITIONS(-Os -ggdb -Wall -Werror --std=gnu99 -Wmissing-declarations)
+ADD_DEFINITIONS(-O2 -ggdb -Wall -Werror --std=gnu99 -Wmissing-declarations)
 ADD_EXECUTABLE(upgraded upgraded.c ../watchdog.c)
 TARGET_LINK_LIBRARIES(upgraded ${ubox})
 INSTALL(TARGETS upgraded
