diff --git a/Modules/CTest.cmake b/Modules/CTest.cmake
index 8109108ec3..8c192dc1c8 100644
--- a/Modules/CTest.cmake
+++ b/Modules/CTest.cmake
@@ -47,7 +47,7 @@ the :variable:`CTEST_USE_LAUNCHERS` variable::
 in the ``CTestConfig.cmake`` file.
 #]=======================================================================]
 
-option(BUILD_TESTING "Build the testing tree." ON)
+option(BUILD_TESTING "Build the testing tree." OFF)
 
 # function to turn generator name into a version string
 # like vs9 or vs10
diff --git a/Tests/Contracts/VTK/Dashboard.cmake.in b/Tests/Contracts/VTK/Dashboard.cmake.in
index ae760bce9f..4b5a6e6839 100644
--- a/Tests/Contracts/VTK/Dashboard.cmake.in
+++ b/Tests/Contracts/VTK/Dashboard.cmake.in
@@ -25,7 +25,7 @@ ctest_empty_binary_directory(${CTEST_BINARY_DIRECTORY})
 
 file(WRITE "${CTEST_BINARY_DIRECTORY}/CMakeCache.txt" "
   VTK_BUILD_EXAMPLES:BOOL=ON
-  VTK_BUILD_TESTING:STRING=WANT
+  VTK_BUILD_TESTING:BOOL=OFF
   VTK_WRAP_PYTHON:BOOL=ON
   ExternalData_OBJECT_STORES:FILEPATH=@base_dir@/ExternalData
 ")
