diff -Naur linux-4.14.95-orig/init/Kconfig linux-4.14.95/init/Kconfig
--- linux-4.14.95-orig/init/Kconfig	2019-01-26 09:05:41.602972835 +0300
+++ linux-4.14.95/init/Kconfig	2019-01-23 10:09:52.000000000 +0300
@@ -1042,6 +1042,13 @@
 	  with the "-O2" compiler flag for best performance and most
 	  helpful compile-time warnings.
 
+config CC_OPTIMIZE_HARDER
+	bool "Optimize harder"
+	help
+	  This option will pass "-O3" to your compiler resulting in a
+	  larger and faster kernel. The more complex optimizations also
+	  increase compilation time and may affect stability.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size"
 	help
diff -Naur linux-4.14.95-orig/Makefile linux-4.14.95/Makefile
--- linux-4.14.95-orig/Makefile	2019-01-26 09:05:43.206306207 +0300
+++ linux-4.14.95/Makefile	2019-01-23 10:09:52.000000000 +0300
@@ -656,12 +656,16 @@
 KBUILD_CFLAGS	+= $(call cc-option,-Oz,-Os)
 KBUILD_CFLAGS	+= $(call cc-disable-warning,maybe-uninitialized,) $(EXTRA_OPTIMIZATION)
 else
+ifdef CONFIG_CC_OPTIMIZE_HARDER
+KBUILD_CFLAGS	+= -O3 $(call cc-disable-warning,maybe-uninitialized,) $(EXTRA_OPTIMIZATION)
+else
 ifdef CONFIG_PROFILE_ALL_BRANCHES
 KBUILD_CFLAGS	+= -O2 $(call cc-disable-warning,maybe-uninitialized,) $(EXTRA_OPTIMIZATION)
 else
 KBUILD_CFLAGS   += -O2 -fno-reorder-blocks -fno-tree-ch $(EXTRA_OPTIMIZATION)
 endif
 endif
+endif
 
 KBUILD_CFLAGS += $(call cc-ifversion, -lt, 0409, \
 			$(call cc-disable-warning,maybe-uninitialized,))
