From 2fec82b1669bb35062ff9029c4b42fb4998fbd44 Mon Sep 17 00:00:00 2001
From: Jim Hague <jim@sinodun.com>
Date: Mon, 2 Mar 2020 13:14:00 +0000
Subject: [PATCH] Revise install of stubby.yml to handle destinations with
 spaces in them.

---
 CMakeLists.txt | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4ba36ed..f3291d7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -117,12 +117,13 @@ install(FILES AUTHORS COPYING ChangeLog NEWS README.md DESTINATION share/doc/stu
 file(GENERATE OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/stubby.yml INPUT ${CMAKE_CURRENT_SOURCE_DIR}/stubby.yml.example)
 # Copy to destination iff no destination file exists.
 install(CODE "\
-  set(targetdir ${STUBBYCONFDIR})\n\
-  if (\$ENV{DESTDIR})\n\
-    set(targetdir \$ENV{DESTDIR}/\${targetdir})\n\
+  set(targetdir \"${STUBBYCONFDIR}\")\n\
+  set(destdir \"\$ENV{DESTDIR}\")\n\
+  if (destdir)\n\
+    set(targetdir \"\${destdir}/\${targetdir}\")\n\
   endif ()\n\
-  if (NOT EXISTS \${targetdir}/stubby.yml)\n\
-    file(COPY ${CMAKE_CURRENT_BINARY_DIR}/stubby.yml DESTINATION \${targetdir})\n\
+  if (NOT EXISTS \"\${targetdir}/stubby.yml\")\n\
+    file(COPY \"${CMAKE_CURRENT_BINARY_DIR}/stubby.yml\" DESTINATION \"\${targetdir}\")\n\
   endif ()")
 
 if (APPLE)
