From c4841747a8042dfe2ea43c3c58d07a87e5a5a0ed Mon Sep 17 00:00:00 2001
From: Jim Hague <jim@sinodun.com>
Date: Wed, 4 Mar 2020 15:08:55 +0000
Subject: [PATCH] Further revise install of stubby.yml to correctly handle
 DESTDIR if present.

---
 CMakeLists.txt | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index f3291d7..92fc1c7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -115,15 +115,25 @@ install(FILES AUTHORS COPYING ChangeLog NEWS README.md DESTINATION share/doc/stu
 
 # Ensure the file gets CRLF line endings on Windows.
 file(GENERATE OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/stubby.yml INPUT ${CMAKE_CURRENT_SOURCE_DIR}/stubby.yml.example)
-# Copy to destination iff no destination file exists.
+# Copy stubby.yml to destination iff no destination file exists.
+# This is complicated by (a) not being able to use generators, due to
+# CMake minimum version requirement, and (b) to account for DESTDIR.
+# And in the latter case, if we're adding DESTDIR to the start of the
+# path, we must on Windows remove any initial drive letter. That's what
+# INSTALL appears to do.
 install(CODE "\
   set(targetdir \"${STUBBYCONFDIR}\")\n\
   set(destdir \"\$ENV{DESTDIR}\")\n\
   if (destdir)\n\
-    set(targetdir \"\${destdir}/\${targetdir}\")\n\
+    string(REGEX REPLACE \"^([A-Z]:)?/(.*)\" \"\\\\2\" newtarget \"\${targetdir}\")\n\
+    if (newtarget)\n\
+      set(targetdir \"\${newtarget}\")\n\
+    endif ()\n\
+    set(targetdir \"\${destdir}/\${newtarget}\")\n\
   endif ()\n\
   if (NOT EXISTS \"\${targetdir}/stubby.yml\")\n\
     file(COPY \"${CMAKE_CURRENT_BINARY_DIR}/stubby.yml\" DESTINATION \"\${targetdir}\")\n\
+    message(\"-- Installing: \${targetdir}/stubby.yml\")\n\
   endif ()")
 
 if (APPLE)
