diff -Naur telegram-cli-1.3.1-20160323-orig/tgl/mtproto-utils.c telegram-cli-1.3.1-20160323/tgl/mtproto-utils.c
--- telegram-cli-1.3.1-20160323-orig/tgl/mtproto-utils.c	2018-02-21 04:53:24.628861373 +0300
+++ telegram-cli-1.3.1-20160323/tgl/mtproto-utils.c	2016-03-23 14:42:53.000000000 +0300
@@ -98,7 +98,7 @@
   if (sizeof (unsigned long) == 8) {
     return TGLC_bn_get_word (b);
   } else if (sizeof (unsigned long long) == 8) {
-    assert (0); // As long as nobody ever uses this code, assume it is broken.
+    // assert (0); // As long as nobody ever uses this code, assume it is broken.
     unsigned long long tmp;
     /* Here be dragons, but it should be okay due to be64toh */
     TGLC_bn_bn2bin (b, (unsigned char *) &tmp);
@@ -112,7 +112,7 @@
   if (sizeof (unsigned long) == 8 || val < (1ll << 32)) {
     TGLC_bn_set_word (b, val);
   } else if (sizeof (unsigned long long) == 8) {
-    assert (0); // As long as nobody ever uses this code, assume it is broken.
+    // assert (0); // As long as nobody ever uses this code, assume it is broken.
     htobe64(val);
     /* Here be dragons, but it should be okay due to htobe64 */
     TGLC_bn_bin2bn ((unsigned char *) &val, 8, b);
