--- bison-3.7.2-orig/lib/stdio.in.h	2020-09-08 10:25:32.739828193 +0300
+++ bison-3.7.2/lib/stdio.in.h	2020-09-08 10:26:52.754942086 +0300
@@ -835,14 +835,6 @@
 # endif
 #endif
 
-/* It is very rare that the developer ever has full control of stdin,
-   so any use of gets warrants an unconditional warning; besides, C11
-   removed it.  */
-#undef gets
-#if HAVE_RAW_DECL_GETS && !defined __cplusplus
-_GL_WARN_ON_USE (gets, "gets is a security hole - use fgets instead");
-#endif
-
 #if defined _WIN32 && !defined __CYGWIN__
 # undef getw
 # define getw _getw
