diff --git a/configure.ac b/configure.ac
index cb81095..cc78212 100644
--- a/configure.ac
+++ b/configure.ac
@@ -197,8 +197,8 @@ if test x"${enable_static}" != x"no"; then
         ;;
     esac
 
-    AX_CHECK_LINK_FLAG([-static-libgcc], [STATIC_LDFLAGS="$STATIC_LDFLAGS -static-libgcc"])
-
+#    AX_CHECK_LINK_FLAG([-static-libgcc], [STATIC_LDFLAGS="$STATIC_LDFLAGS -static-libgcc"])
+#
     STATIC_LDFLAGS="$STATIC_LDFLAGS -L`pwd`/src/static-libs"
     AX_CHECK_LINK_FLAG([-static-libstdc++],
                        [STATIC_LDFLAGS="$STATIC_LDFLAGS -static-libstdc++"])
