diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..6abeb40
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,27 @@
+cmake_minimum_required(VERSION 3.0.0)
+project(polipo VERSION 0.1.0)
+
+include(CTest)
+enable_testing()
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11")
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Os -g -Wall -fno-strict-aliasing")
+
+set(HEADERS
+atom.h auth.h chunk.h client.h config.h dirent_compat.h diskcache.h
+dns.h event.h forbidden.h http_parse.h http.h
+io.h local.h log.h md5.h mingw.h object.h parse_time.h polipo.h
+server.h socks.h tunnel.h util.h
+)
+
+set(SRCS  
+util.c event.c io.c chunk.c atom.c object.c log.c diskcache.c main.c 
+config.c local.c http.c client.c server.c auth.c tunnel.c 
+http_parse.c parse_time.c dns.c forbidden.c 
+md5.c socks.c mingw.c)
+
+add_executable(polipo ${SRCS} ${HEADERS})
+
+set(CPACK_PROJECT_NAME ${PROJECT_NAME})
+set(CPACK_PROJECT_VERSION ${PROJECT_VERSION})
+include(CPack)
\ No newline at end of file
diff --git a/diskcache.c b/diskcache.c
index 270d67a..c91134b 100644
--- a/diskcache.c
+++ b/diskcache.c
@@ -24,7 +24,7 @@ THE SOFTWARE.
 
 #ifndef NO_DISK_CACHE
 
-#include "md5import.h"
+#include "md5.h"
 
 int maxDiskEntries = 32;
 
diff --git a/md5.h b/md5.h
index aaf9548..86dfddf 100644
--- a/md5.h
+++ b/md5.h
@@ -32,6 +32,14 @@
  ***********************************************************************
  */
 
+#if defined __STDC_VERSION__ && __STDC_VERSION__ >= 199901L
+#define HAS_STDINT_H
+#else
+typedef unsigned int my_uint32_t;
+#undef uint32_t
+#define uint32_t my_uint32_t
+#endif
+
 #ifdef HAS_STDINT_H
 #include <stdint.h>
 #elif defined(HAS_INTTYPES_H)
diff --git a/md5import.c b/md5import.c
deleted file mode 100644
index 2e74c7b..0000000
--- a/md5import.c
+++ /dev/null
@@ -1,15 +0,0 @@
-#ifndef _GNU_SOURCE
-#define _GNU_SOURCE
-#endif
-
-#include <stdlib.h>
-
-#if defined __STDC_VERSION__ && __STDC_VERSION__ >= 199901L
-#define HAS_STDINT_H
-#else
-typedef unsigned int my_uint32_t;
-#undef uint32_t
-#define uint32_t my_uint32_t
-#endif
-#include "md5.c"
-#undef uint32_t
diff --git a/md5import.h b/md5import.h
deleted file mode 100644
index c97c632..0000000
--- a/md5import.h
+++ /dev/null
@@ -1,9 +0,0 @@
-#if defined __STDC_VERSION__ && __STDC_VERSION__ >= 199901L
-#define HAS_STDINT_H
-#else
-typedef unsigned int my_uint32_t;
-#undef uint32_t
-#define uint32_t my_uint32_t
-#endif
-#include "md5.h"
-#undef uint32_t
